<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load Hap.py Results into R • happyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">happyR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="/index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Precision-recall_curves.html">Precision-recall curves</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    

    
    
<div class="contents">
<div id="happyr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#happyr" class="anchor"></a>happyR</h1></div>

<p>Load <a href="https://github.com/Illumina/hap.py">hap.py</a> results into an R data structure to enable simple plotting, comparisons and aggregation. The related package <a href="https://github.com/Illumina/happyCompare">happyCompare</a> builds on happyR with statistical analysis of groups and replicates, as well as providing some more sophisticated plotting and reporting functionality.</p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Use devtools:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">"Illumina/happyR"</span>)</code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(happyR)

<span class="co"># demo data that comes with the package</span>
happy_input &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"happy_demo.summary.csv"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)
happy_prefix &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">".summary.csv"</span>, <span class="st">""</span>, happy_input)

<span class="co"># happy_prefix is the -o argument to hap.py, here: path/to/files/happy_demo</span>
hapdata &lt;-<span class="st"> </span><span class="kw"><a href="reference/read_happy.html">read_happy</a></span>(happy_prefix)
<span class="co">#&gt; Reading summary table</span>
<span class="co">#&gt; Reading extended table</span>
<span class="co">#&gt; Reading precision-recall curve data</span>
hapdata
<span class="co">#&gt;   Hap.py result containing:  summary, extended, pr_curve </span>
<span class="co">#&gt;    Loaded from:  /Users/bmoore1/Rlibs/happyR/extdata/happy_demo  (hap.py version: v0.3.9)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; # A tibble: 4 x 17</span>
<span class="co">#&gt;    Type Filter TRUTH.TOTAL TRUTH.TP TRUTH.FN QUERY.TOTAL QUERY.FP</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;       &lt;int&gt;    &lt;int&gt;    &lt;int&gt;       &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1 INDEL    ALL        8937     7839     1098       11812      343</span>
<span class="co">#&gt; 2 INDEL   PASS        8937     7550     1387        9971      283</span>
<span class="co">#&gt; 3   SNP    ALL       52494    52125      369       90092      582</span>
<span class="co">#&gt; 4   SNP   PASS       52494    46920     5574       48078      143</span>
<span class="co">#&gt; # ... with 10 more variables: QUERY.UNK &lt;int&gt;, FP.gt &lt;int&gt;,</span>
<span class="co">#&gt; #   METRIC.Recall &lt;dbl&gt;, METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;,</span>
<span class="co">#&gt; #   METRIC.F1_Score &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span>
<span class="co">#&gt; #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;,</span>
<span class="co">#&gt; #   QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;</span></code></pre></div>
<p><code>hapdata</code> is now a single hap.py data object containing:</p>
<ul>
<li>
<code>summary</code> (from summary.csv) - high-level ALL / PASS numbers</li>
<li>
<code>extended</code> (from extended.csv) - region / subtype stratified metrics</li>
<li>
<code>pr_curve</code> (from roc.*.csv.gz) - precision-recall over quality score</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(hapdata)
<span class="co">#&gt; [1] "summary"  "extended" "pr_curve"</span>

<span class="co"># e.g. here pr_curve$INDEL_PASS maps to happy_demo.roc.Locations.INDEL.PASS.csv.gz</span>
<span class="kw">names</span>(hapdata$pr_curve)
<span class="co">#&gt; [1] "INDEL_SEL"  "INDEL_PASS" "SNP"        "all"        "INDEL"     </span>
<span class="co">#&gt; [6] "SNP_PASS"   "SNP_SEL"</span></code></pre></div>
</div>
<div id="example-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#example-plots" class="anchor"></a>Example plots</h2>
<div id="indel-subtypes" class="section level3">
<h3 class="hasAnchor">
<a href="#indel-subtypes" class="anchor"></a>Indel subtypes</h3>
<p>Here subtypes are categories based on the variant type and length (e.g. <code>I6_15</code> are insertions of length 6 bp up to 15 bp). Genomic subsets are <code>*</code> for all, <code>TS_contained</code> for truth variants fully-contained with confident regions and <code>TS_boundary</code> for truth variants near the edge of a confident region block. See <a href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md">hap.py docs</a> for more info.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

<span class="co"># get indel subtypes from 'extended', skipping complex alleles and combined</span>
indel_extended &lt;-<span class="st"> </span><span class="kw">subset</span>(hapdata$extended, Type ==<span class="st"> "INDEL"</span> &amp;<span class="st"> </span>
<span class="st">                         </span>Filter ==<span class="st"> "ALL"</span> &amp;<span class="st"> </span><span class="kw">grepl</span>(<span class="st">"^[DI]"</span>, Subtype))

<span class="co"># Precision-recall by subtype, scaled by number in truthset</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(indel_extended, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, 
                           <span class="dt">col =</span> Subtype, <span class="dt">size =</span> TRUTH.TOTAL, <span class="dt">shape=</span>Subset)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>() +<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>() +<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_brewer">scale_color_brewer</a></span>(<span class="dt">palette =</span> <span class="st">"Set2"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_size">scale_size</a></span>(<span class="dt">guide =</span> <span class="st">"none"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Indel subtype precision and recall"</span>)</code></pre></div>
<p><img src="examples/README-indel_subtypes-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="precision-recall-curves" class="section level3">
<h3 class="hasAnchor">
<a href="#precision-recall-curves" class="anchor"></a>Precision-recall curves</h3>
<p>PR curves show how precision and recall vary with a changing threshold, in this example it’s over a range of quality score thresholds applied to a set of variants: as the threshold increases, the remaining variant set is less comprehensive (lower recall) but typically contains fewer false positives (higher precision).</p>
<p>This gets more complicated when <code>PASS</code> records aren’t set solely by a single quality score threshold, but by multiple independent filters (e.g. high-depth, genomic context, etc.). These interactions can’t be fully captured by drawing a PR curve only on a quality threshold:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using happyR::pr_data to simplify subsetting:</span>
all_pr &lt;-<span class="st"> </span><span class="kw"><a href="reference/pr_data.html">pr_data</a></span>(hapdata)

<span class="co"># this gets PR curve starting at ALL point, equivalent to base:</span>
<span class="co"># all_pr &lt;- subset(hapdata$pr_curve$all, Filter == "ALL" &amp; Subtype == "*" &amp; Subset == "*")</span>

<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(all_pr, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Type)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>(<span class="dt">data =</span> hapdata$summary) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_x_continuous</a></span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">6</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_y_continuous</a></span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">95</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"ALL PR curve might not hit the PASS point"</span>)</code></pre></div>
<p><img src="examples/README-all_pr-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Hap.py accounts for this by generating a selectively-filtered PR curve (<code>SEL</code>). First these independent filters are applied to the variant set, then we can draw the remaining PR curve using the quality score threshold. This gives a more accurate view of how changing the quality score threshold can impact precision and recall.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># selectively filtered PR curve</span>
pr &lt;-<span class="st"> </span><span class="kw"><a href="reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">filter =</span> <span class="st">"SEL"</span>)

<span class="co"># link this to the ALL point</span>
pr &lt;-<span class="st"> </span>dplyr::<span class="kw">bind_rows</span>(pr, <span class="kw">subset</span>(hapdata$summary, Filter ==<span class="st"> "ALL"</span>))

<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(pr, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Type)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>(<span class="dt">data =</span> hapdata$summary) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_x_continuous</a></span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">6</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous">scale_y_continuous</a></span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">95</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Selectively-filtered PR curve reliably hits PASS"</span>)</code></pre></div>
<p><img src="examples/README-sel_pr-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="system-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#system-requirements" class="anchor"></a>System requirements</h2>
<p>Originally developed for R v3.4.0. <a href="https://travis-ci.org/Illumina/happyR">Tests</a> are run using the most recent available R versions (incl. devel) on Ubuntu (Trusty) and OS X (El Capitan) platforms. HappyR has not been tested on Windows. Dependencies are listed in <a href="DESCRIPTION" class="uri">DESCRIPTION</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2 class="hasAnchor">
<a href="#sidebar" class="anchor"></a>License</h2>
<p>file <a href="LICENSE.html">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ben Moore <br><small class="roles"> Author, maintainer </small> </li>
<li>Mar Gonzalez-Porta <br><small class="roles"> Author </small> </li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/Illumina/happyR"><img src="https://travis-ci.org/Illumina/happyR.svg?branch=master" alt="Build Status"></a></li>
</ul>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
