<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load hap.py results into R â€¢ happyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Load hap.py results into R">
<meta property="og:description" content="Read multiple output files from the haplotype-comparison
    tool hap.py into an organised R data structure.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">happyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Precision-recall_curves.html">Precision-recall curves</a>
    </li>
    <li>
      <a href="articles/loading_data.html">Loading hap.py results</a>
    </li>
    <li>
      <a href="articles/stratified_counts.html">Analysing stratified counts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<div id="happyr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#happyr" class="anchor"></a>happyR</h1></div>

<p>Tools to help analyse your <a href="https://github.com/Illumina/hap.py">hap.py</a> results in R. See the <a href="https://illumina.github.io/happyR">documentation</a> for usage and examples.</p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"Illumina/happyR"</span>)</a></code></pre></div>
</div>
<div id="demo" class="section level2">
<h2 class="hasAnchor">
<a href="#demo" class="anchor"></a>Demo</h2>
<p>This example walks through a comparison of samples prepared using PCR-Free versus Nano library preps with 2 replicates per group.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(happyR)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co"># groups are defined either by a CSV or data.frame with three </span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># required columns: a label for each group (group_id), a unique</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># label per replicate (replicate_id) and a path to the respective</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co"># pre-computed hap.py output (happy_prefix)</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">extdata_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co"># these extdata files are supplied with the package</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10">samplesheet &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">  <span class="op">~</span>group_id,  <span class="op">~</span>replicate_id, <span class="op">~</span>happy_prefix,</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">  <span class="st">"PCR-Free"</span>, <span class="st">"NA12878-I30"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(extdata_dir, <span class="st">"NA12878-I30_S1"</span>, <span class="dt">sep =</span> <span class="st">"/"</span>),</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">  <span class="st">"PCR-Free"</span>, <span class="st">"NA12878-I33"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(extdata_dir, <span class="st">"NA12878-I33_S1"</span>, <span class="dt">sep =</span> <span class="st">"/"</span>),</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">  <span class="st">"Nano"</span>,     <span class="st">"NA12878-R1"</span>,  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(extdata_dir, <span class="st">"NA12878-R1_S1"</span>, <span class="dt">sep =</span> <span class="st">"/"</span>),</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">  <span class="st">"Nano"</span>,     <span class="st">"NA12878-R2"</span>,  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(extdata_dir, <span class="st">"NA12878-R2_S1"</span>, <span class="dt">sep =</span> <span class="st">"/"</span>)</a>
<a class="sourceLine" id="cb2-16" data-line-number="16">)</a>
<a class="sourceLine" id="cb2-17" data-line-number="17"></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co"># here the above table is used to read hap.py output files from disk </span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co"># and attach the group + replicate metadata</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20">hap_samplesheet &lt;-<span class="st"> </span><span class="kw"><a href="reference/read_samplesheet_.html">read_samplesheet_</a></span>(samplesheet)</a>
<a class="sourceLine" id="cb2-21" data-line-number="21"></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co"># extract summary PASS performance for each replicate and plot</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23">summary &lt;-<span class="st"> </span><span class="kw"><a href="reference/extract_results.html">extract_results</a></span>(hap_samplesheet<span class="op">$</span>results, <span class="dt">table =</span> <span class="st">"summary"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="st">  </span><span class="kw">inner_join</span>(samplesheet, <span class="dt">by =</span> <span class="st">"happy_prefix"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Filter <span class="op">==</span><span class="st"> "PASS"</span>)</a>
<a class="sourceLine" id="cb2-26" data-line-number="26"></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="kw">ggplot</span>(<span class="dt">data =</span> summary, <span class="kw">aes</span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">color =</span> group_id, <span class="dt">shape =</span> Type)) <span class="op">+</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">xlim</a></span>(<span class="ot">NA</span>, <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">ylim</a></span>(<span class="ot">NA</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="st">  </span><span class="kw">scale_color_brewer</span>(<span class="dt">palette =</span> <span class="st">"Set2"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"Recall"</span>, <span class="dt">y =</span> <span class="st">"Precision"</span>, <span class="dt">color =</span> <span class="st">"Prep"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"PCR-Free vs. Nano variant calling"</span>,</a>
<a class="sourceLine" id="cb2-33" data-line-number="33">          <span class="st">"PCR treatment reduces indel calling performance"</span>)  </a></code></pre></div>
<p><img src="examples/README-usage-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="system-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#system-requirements" class="anchor"></a>System requirements</h2>
<p>Originally developed for R v3.4.0. <a href="https://travis-ci.org/Illumina/happyR">Tests</a> are run using the most recent available R versions (incl. devel) on Ubuntu (Trusty) and OS X (El Capitan) platforms. happyR has not been tested on Windows. Dependencies are listed in <a href="DESCRIPTION">DESCRIPTION</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ben Moore <br><small class="roles"> Author, maintainer </small>  </li>
<li>Mar Gonzalez-Porta <br><small class="roles"> Author </small>  </li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/Illumina/happyR"><img src="https://travis-ci.org/Illumina/happyR.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/gh/Illumina/happyR"><img src="https://codecov.io/gh/Illumina/happyR/branch/master/graph/badge.svg" alt="codecov"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
