<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load Hap.py Results into R • happyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">happyR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="happyr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#happyr" class="anchor"> </a>happyR</h1></div>

<p>Load hap.py results into a useful R data structure</p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"> </a>Install</h2>
<p>Download <a href="https://git.illumina.com/bmoore1/happyR/archive/master.zip">zip</a>, extract and run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_local</span>(<span class="st">"path/to/happyR-master/"</span>)</code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"> </a>Usage</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(happyR)

<span class="co"># demo data that comes with the package</span>
happy_input &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"happy_demo.summary.csv"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)
happy_prefix &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">".summary.csv"</span>, <span class="st">""</span>, happy_input)

<span class="co"># happy_prefix is the -o argument to hap.py, here: path/to/files/happy_demo</span>
hapdata &lt;-<span class="st"> </span><span class="kw"><a href="reference/read_happy.html">read_happy</a></span>(happy_prefix)
hapdata
<span class="co"># Hap.py result containing:  summary, extended, pr_curve </span>
<span class="co"># </span>
<span class="co"># # happy_summary [4 × 16]</span>
<span class="co"># Type Filter TRUTH.TOTAL TRUTH.TP TRUTH.FN QUERY.TOTAL QUERY.FP QUERY.UNK FP.gt METRIC.Recall</span>
<span class="co"># &lt;chr&gt;  &lt;chr&gt;       &lt;int&gt;    &lt;int&gt;    &lt;int&gt;       &lt;int&gt;    &lt;int&gt;     &lt;int&gt; &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co"># 1 INDEL    ALL        8946     7840     1106       11811      367      3494    46      0.876369</span>
<span class="co"># 2 INDEL   PASS        8946     7551     1395        9970      301      1944    31      0.844064</span>
<span class="co"># 3   SNP    ALL       52494    52125      369       90092      548     37383   107      0.992971</span>
<span class="co"># 4   SNP   PASS       52494    46920     5574       48078      122      1014     8      0.893816</span>
<span class="co"># # ... with 6 more variables: METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span>
<span class="co"># #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;</span></code></pre></div>
<p><code>hapdata</code> is now a single hap.py data object containing:</p>
<ul>
<li>
<code>summary</code> (from summary.csv) - high-level ALL / PASS numbers</li>
<li>
<code>extended</code> (from extended.csv) - region / subtype stratified metrics</li>
<li>
<code>pr_curve</code> (from roc.*.csv.gz) - precision-recall over quality score</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(hapdata)
<span class="co"># [1] "summary"  "extended" "pr_curve"</span>

<span class="co"># e.g. here pr_curve$INDEL_PASS maps to happy_demo.roc.Locations.INDEL.PASS.csv.gz</span>
<span class="kw">names</span>(hapdata$pr_curve)
<span class="co"># [1] "all"        "INDEL"      "INDEL_PASS" "INDEL_SEL"  "SNP"        "SNP_PASS"   "SNP_SEL"   </span></code></pre></div>
</div>
<div id="example-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#example-plots" class="anchor"> </a>Example plots</h2>
<div id="indel-subtypes" class="section level3">
<h3 class="hasAnchor">
<a href="#indel-subtypes" class="anchor"> </a>Indel subtypes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get indel subtypes from 'extended', skipping complex alleles and combined
indel_extended &lt;-<span class="st"> </span><span class="kw">subset</span>(hapdata$extended, Type ==<span class="st"> "INDEL"</span> &amp;<span class="st"> </span>Filter ==<span class="st"> "ALL"</span> &amp;<span class="st"> </span><span class="kw">grepl</span>(<span class="st">"^[DI]"</span>, Subtype))

<span class="co"># Precision-recall by subtype, scaled by number in truthset</span>
<span class="kw">ggplot</span>(indel_extended, <span class="kw">aes</span>(<span class="dt">x=</span>METRIC.Recall, <span class="dt">y=</span>METRIC.Precision, <span class="dt">col=</span>Subtype, <span class="dt">size=</span>TRUTH.TOTAL)) +
<span class="st">  </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">theme_minimal</span>() +<span class="st"> </span>
<span class="st">  </span><span class="kw">scale_color_brewer</span>(<span class="dt">palette =</span> <span class="st">"Set2"</span>) +
<span class="st">  </span><span class="kw">scale_size</span>(<span class="dt">guide =</span> <span class="st">"none"</span>) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Indel subtype precision and recall"</span>)</code></pre></div>
<p><img src="https://git.illumina.com/bmoore1/happyR/raw/master/examples/happyr_eg_indelext.png" width="600"></p>
</div>
<div id="precision-recall-curves" class="section level3">
<h3 class="hasAnchor">
<a href="#precision-recall-curves" class="anchor"> </a>Precision-recall curves</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># PR curve starting at ALL point</span>
all_pr &lt;-<span class="st"> </span><span class="kw">subset</span>(hapdata$pr_curve$all, Filter ==<span class="st"> "ALL"</span> &amp;<span class="st"> </span>Subtype ==<span class="st"> "*"</span>)

<span class="kw">ggplot</span>(all_pr, <span class="kw">aes</span>(<span class="dt">x=</span>METRIC.Recall, <span class="dt">y=</span>METRIC.Precision, <span class="dt">col=</span>Type)) +
<span class="st">  </span><span class="kw">geom_line</span>() +<span class="st"> </span><span class="kw">theme_minimal</span>() +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data=</span>hapdata$summary) +
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">6</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">95</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"ALL PR curve might not hit the PASS point"</span>)</code></pre></div>
<p><img src="https://git.illumina.com/bmoore1/happyR/raw/master/examples/happyr_eg_prall.png" width="600"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># selectively filtered PR curve</span>
pr &lt;-<span class="st"> </span><span class="kw">subset</span>(hapdata$pr_curve$all, Filter ==<span class="st"> "SEL"</span> &amp;<span class="st"> </span>Subtype ==<span class="st"> "*"</span>)

<span class="co"># link this to the ALL point</span>
pr &lt;-<span class="st"> </span>dplyr::<span class="kw">bind_rows</span>(pr, <span class="kw">subset</span>(hapdata$summary, Filter ==<span class="st"> "ALL"</span>))

<span class="kw">ggplot</span>(pr, <span class="kw">aes</span>(<span class="dt">x=</span>METRIC.Recall, <span class="dt">y=</span>METRIC.Precision, <span class="dt">col=</span>Type)) +
<span class="st">  </span><span class="kw">geom_line</span>() +<span class="st"> </span><span class="kw">theme_minimal</span>() +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data=</span>hapdata$summary) +
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">6</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="dt">limits =</span> <span class="kw">c</span>(.<span class="dv">95</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Selectively-filtered PR curve reliably hits PASS"</span>)</code></pre></div>
<p><img src="https://git.illumina.com/bmoore1/happyR/raw/master/examples/happyr_eg_prsel.png" width="600"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ben Moore <br><small class="roles"> Author, maintainer </small> </li>
<li>Mar Gonzalez-Porta <br><small class="roles"> Author </small> </li>
</ul>
<html><body>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://ukch-prd-jnks01.illumina.com/job/happyR/job/master/"><img src="https://ukch-prd-jnks01.illumina.com/buildStatus/icon?job=happyR/master" alt="Build Status"></a></li>
</ul>
</body></html>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
