<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading hap.py results • happyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Loading hap.py results">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">happyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/loading_data.html">Loading hap.py results</a>
    </li>
    <li>
      <a href="../articles/precision-recall_curves.html">Precision-recall curves</a>
    </li>
    <li>
      <a href="../articles/stratified_counts.html">Analysing stratified counts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Loading hap.py results</h1>
                        <h4 class="author">Mar Gonzalez-Porta</h4>
            
            <h4 class="date">2019-06-26</h4>
      
      
      <div class="hidden name"><code>loading_data.Rmd</code></div>

    </div>

    
    
<p>happyR provides an easy framework to import hap.py results into R. The user can decide whether to upload results from a single sample, by providing a happy prefix path, or to bulk load multiple datasets via a samplesheet. In both cases, it is easy to extract relevant results using helper functions.</p>
<div id="set-up" class="section level2">
<h2 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set up</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(happyR)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)</a></code></pre></div>
</div>
<div id="loading-and-querying-individual-hap-py-results" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-and-querying-individual-hap-py-results" class="anchor"></a>Loading and querying individual hap.py results</h2>
<p>Let’s first explore how to load hap.py outputs from a single sample:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># define happy_prefix (the -o argument to hap.py, here: path/to/files/happy_demo)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">extdata_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">happy_prefix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"%s/happy_demo"</span>, extdata_dir)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># load hap.py results</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">hap_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_happy.html">read_happy</a></span>(happy_prefix)</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; Reading summary table</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; Reading extended table</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt; Reading precision-recall curve data</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(hap_result)</a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt; [1] "happy_result"</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(hap_result)</a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt; [1] "summary"  "extended" "pr_curve"</span></a></code></pre></div>
<p><code>hap_result</code> is now a <code>happy_result</code> object with the following fields:</p>
<ul>
<li>
<code>summary</code> (from <code>summary.csv</code>) - a data.frame with high-level ALL / PASS numbers</li>
<li>
<code>extended</code> (from <code>extended.csv</code>) - a data.frame with region / subtype stratified metrics</li>
<li>
<code>pr_curve</code> (from <code>roc.*.csv.gz</code>) - a <code>happy_roc</code> object (a list of data.frames) containing precision-recall over quality score</li>
</ul>
<p>We can query each item in our <code>happy_result</code> object using standard R syntax:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">hap_result<span class="op">$</span>summary <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">#&gt; # A tibble: 4 x 17</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt;   Type  Filter TRUTH.TOTAL TRUTH.TP TRUTH.FN QUERY.TOTAL QUERY.FP QUERY.UNK</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; 1 INDEL ALL           8937     7839     1098       11812      343      3520</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; 2 INDEL PASS          8937     7550     1387        9971      283      1964</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; 3 SNP   ALL          52494    52125      369       90092      582     37348</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; 4 SNP   PASS         52494    46920     5574       48078      143       992</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; # … with 9 more variables: FP.gt &lt;dbl&gt;, METRIC.Recall &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; #   METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;, METRIC.F1_Score &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt; #   TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"></a>
<a class="sourceLine" id="cb3-14" data-line-number="14">hap_result<span class="op">$</span>extended <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co">#&gt; # A tibble: 6 x 65</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">#&gt;   Type  Subtype Subset Filter Genotype QQ.Field    QQ METRIC.Recall</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;    &lt;dbl&gt;         &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">#&gt; 1 INDEL *       *      ALL    *        QUAL        NA         0.877</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="co">#&gt; 2 INDEL *       *      PASS   *        QUAL        NA         0.845</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20"><span class="co">#&gt; 3 INDEL *       TS_bo… ALL    *        QUAL        NA         0.527</span></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="co">#&gt; 4 INDEL *       TS_bo… PASS   *        QUAL        NA         0.438</span></a>
<a class="sourceLine" id="cb3-22" data-line-number="22"><span class="co">#&gt; 5 INDEL *       TS_co… ALL    *        QUAL        NA         0.914</span></a>
<a class="sourceLine" id="cb3-23" data-line-number="23"><span class="co">#&gt; 6 INDEL *       TS_co… PASS   *        QUAL        NA         0.887</span></a>
<a class="sourceLine" id="cb3-24" data-line-number="24"><span class="co">#&gt; # … with 57 more variables: METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-25" data-line-number="25"><span class="co">#&gt; #   METRIC.F1_Score &lt;dbl&gt;, FP.gt &lt;int&gt;, FP.al &lt;int&gt;, Subset.Size &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-26" data-line-number="26"><span class="co">#&gt; #   Subset.IS_CONF.Size &lt;dbl&gt;, Subset.Level &lt;dbl&gt;, TRUTH.TOTAL &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb3-27" data-line-number="27"><span class="co">#&gt; #   TRUTH.TOTAL.ti &lt;dbl&gt;, TRUTH.TOTAL.tv &lt;dbl&gt;, TRUTH.TOTAL.het &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-28" data-line-number="28"><span class="co">#&gt; #   TRUTH.TOTAL.homalt &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-29" data-line-number="29"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, TRUTH.TP &lt;int&gt;, TRUTH.TP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="co">#&gt; #   TRUTH.TP.tv &lt;dbl&gt;, TRUTH.TP.het &lt;dbl&gt;, TRUTH.TP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="co">#&gt; #   TRUTH.TP.TiTv_ratio &lt;dbl&gt;, TRUTH.TP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-32" data-line-number="32"><span class="co">#&gt; #   TRUTH.FN &lt;int&gt;, TRUTH.FN.ti &lt;dbl&gt;, TRUTH.FN.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="co">#&gt; #   TRUTH.FN.het &lt;dbl&gt;, TRUTH.FN.homalt &lt;dbl&gt;, TRUTH.FN.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-34" data-line-number="34"><span class="co">#&gt; #   TRUTH.FN.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL &lt;int&gt;, QUERY.TOTAL.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-35" data-line-number="35"><span class="co">#&gt; #   QUERY.TOTAL.tv &lt;dbl&gt;, QUERY.TOTAL.het &lt;dbl&gt;, QUERY.TOTAL.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-36" data-line-number="36"><span class="co">#&gt; #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-37" data-line-number="37"><span class="co">#&gt; #   QUERY.TP &lt;int&gt;, QUERY.TP.ti &lt;dbl&gt;, QUERY.TP.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-38" data-line-number="38"><span class="co">#&gt; #   QUERY.TP.het &lt;dbl&gt;, QUERY.TP.homalt &lt;dbl&gt;, QUERY.TP.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-39" data-line-number="39"><span class="co">#&gt; #   QUERY.TP.het_hom_ratio &lt;dbl&gt;, QUERY.FP &lt;int&gt;, QUERY.FP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-40" data-line-number="40"><span class="co">#&gt; #   QUERY.FP.tv &lt;dbl&gt;, QUERY.FP.het &lt;dbl&gt;, QUERY.FP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-41" data-line-number="41"><span class="co">#&gt; #   QUERY.FP.TiTv_ratio &lt;dbl&gt;, QUERY.FP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-42" data-line-number="42"><span class="co">#&gt; #   QUERY.UNK &lt;int&gt;, QUERY.UNK.ti &lt;dbl&gt;, QUERY.UNK.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-43" data-line-number="43"><span class="co">#&gt; #   QUERY.UNK.het &lt;dbl&gt;, QUERY.UNK.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-44" data-line-number="44"><span class="co">#&gt; #   QUERY.UNK.TiTv_ratio &lt;dbl&gt;, QUERY.UNK.het_hom_ratio &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb3-45" data-line-number="45"></a>
<a class="sourceLine" id="cb3-46" data-line-number="46"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(hap_result<span class="op">$</span>pr_curve)</a>
<a class="sourceLine" id="cb3-47" data-line-number="47"><span class="co">#&gt; [1] "INDEL_SEL"  "INDEL_PASS" "SNP"        "all"        "INDEL"     </span></a>
<a class="sourceLine" id="cb3-48" data-line-number="48"><span class="co">#&gt; [6] "SNP_PASS"   "SNP_SEL"</span></a>
<a class="sourceLine" id="cb3-49" data-line-number="49"><span class="co"># e.g. here pr_curve$INDEL_PASS maps to happy_demo.roc.Locations.INDEL.PASS.csv.gz</span></a>
<a class="sourceLine" id="cb3-50" data-line-number="50">hap_result<span class="op">$</span>pr_curve<span class="op">$</span>INDEL_PASS <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb3-51" data-line-number="51"><span class="co">#&gt; # A tibble: 6 x 66</span></a>
<a class="sourceLine" id="cb3-52" data-line-number="52"><span class="co">#&gt;   Type  Subtype Subset Filter Genotype QQ.Field     QQ METRIC.Recall</span></a>
<a class="sourceLine" id="cb3-53" data-line-number="53"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;     &lt;dbl&gt;         &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb3-54" data-line-number="54"><span class="co">#&gt; 1 INDEL *       *      PASS   *        QUAL     0.          0.845   </span></a>
<a class="sourceLine" id="cb3-55" data-line-number="55"><span class="co">#&gt; 2 INDEL *       *      PASS   *        QUAL     1.09e1      0.845   </span></a>
<a class="sourceLine" id="cb3-56" data-line-number="56"><span class="co">#&gt; 3 INDEL *       *      PASS   *        QUAL     1.00e3      0.482   </span></a>
<a class="sourceLine" id="cb3-57" data-line-number="57"><span class="co">#&gt; 4 INDEL *       *      PASS   *        QUAL     1.00e3      0.482   </span></a>
<a class="sourceLine" id="cb3-58" data-line-number="58"><span class="co">#&gt; 5 INDEL *       *      PASS   *        QUAL     1.00e4      0.000224</span></a>
<a class="sourceLine" id="cb3-59" data-line-number="59"><span class="co">#&gt; 6 INDEL *       *      PASS   *        QUAL     1.00e3      0.482   </span></a>
<a class="sourceLine" id="cb3-60" data-line-number="60"><span class="co">#&gt; # … with 58 more variables: METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-61" data-line-number="61"><span class="co">#&gt; #   METRIC.F1_Score &lt;dbl&gt;, FP.gt &lt;int&gt;, FP.al &lt;int&gt;, Subset.Size &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb3-62" data-line-number="62"><span class="co">#&gt; #   Subset.IS_CONF.Size &lt;dbl&gt;, Subset.Level &lt;dbl&gt;, TRUTH.TOTAL &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb3-63" data-line-number="63"><span class="co">#&gt; #   TRUTH.TOTAL.ti &lt;dbl&gt;, TRUTH.TOTAL.tv &lt;dbl&gt;, TRUTH.TOTAL.het &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-64" data-line-number="64"><span class="co">#&gt; #   TRUTH.TOTAL.homalt &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-65" data-line-number="65"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, TRUTH.TP &lt;int&gt;, TRUTH.TP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-66" data-line-number="66"><span class="co">#&gt; #   TRUTH.TP.tv &lt;dbl&gt;, TRUTH.TP.het &lt;dbl&gt;, TRUTH.TP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-67" data-line-number="67"><span class="co">#&gt; #   TRUTH.TP.TiTv_ratio &lt;dbl&gt;, TRUTH.TP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-68" data-line-number="68"><span class="co">#&gt; #   TRUTH.FN &lt;int&gt;, TRUTH.FN.ti &lt;dbl&gt;, TRUTH.FN.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-69" data-line-number="69"><span class="co">#&gt; #   TRUTH.FN.het &lt;dbl&gt;, TRUTH.FN.homalt &lt;dbl&gt;, TRUTH.FN.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-70" data-line-number="70"><span class="co">#&gt; #   TRUTH.FN.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL &lt;int&gt;, QUERY.TOTAL.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-71" data-line-number="71"><span class="co">#&gt; #   QUERY.TOTAL.tv &lt;dbl&gt;, QUERY.TOTAL.het &lt;dbl&gt;, QUERY.TOTAL.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-72" data-line-number="72"><span class="co">#&gt; #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-73" data-line-number="73"><span class="co">#&gt; #   QUERY.TP &lt;int&gt;, QUERY.TP.ti &lt;dbl&gt;, QUERY.TP.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-74" data-line-number="74"><span class="co">#&gt; #   QUERY.TP.het &lt;dbl&gt;, QUERY.TP.homalt &lt;dbl&gt;, QUERY.TP.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-75" data-line-number="75"><span class="co">#&gt; #   QUERY.TP.het_hom_ratio &lt;dbl&gt;, QUERY.FP &lt;int&gt;, QUERY.FP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-76" data-line-number="76"><span class="co">#&gt; #   QUERY.FP.tv &lt;dbl&gt;, QUERY.FP.het &lt;dbl&gt;, QUERY.FP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-77" data-line-number="77"><span class="co">#&gt; #   QUERY.FP.TiTv_ratio &lt;dbl&gt;, QUERY.FP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-78" data-line-number="78"><span class="co">#&gt; #   QUERY.UNK &lt;int&gt;, QUERY.UNK.ti &lt;dbl&gt;, QUERY.UNK.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-79" data-line-number="79"><span class="co">#&gt; #   QUERY.UNK.het &lt;dbl&gt;, QUERY.UNK.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-80" data-line-number="80"><span class="co">#&gt; #   QUERY.UNK.TiTv_ratio &lt;dbl&gt;, QUERY.UNK.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb3-81" data-line-number="81"><span class="co">#&gt; #   file_source &lt;chr&gt;</span></a></code></pre></div>
<p>Or we can use the helper <code>pr_data</code> function to query <code>pr_curve</code> with advanced filters:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">del_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hap_result, <span class="dt">var_type =</span> <span class="st">"indel"</span>, <span class="dt">filter =</span> <span class="st">"PASS"</span>, <span class="dt">subtype =</span> <span class="st">"*"</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">del_pr <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">#&gt; # A tibble: 6 x 66</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt;   Type  Subtype Subset Filter Genotype QQ.Field     QQ METRIC.Recall</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;     &lt;dbl&gt;         &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; 1 SNP   *       *      PASS   *        QUAL        0           0.894</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; 2 SNP   *       *      PASS   *        QUAL       10.2         0.894</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; 3 SNP   *       *      PASS   *        QUAL       10.8         0.894</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; 4 SNP   *       *      PASS   *        QUAL      100.          0.894</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; 5 SNP   *       *      PASS   *        QUAL      101.          0.894</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; 6 SNP   *       *      PASS   *        QUAL     1000.          0.442</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; # … with 58 more variables: METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt; #   METRIC.F1_Score &lt;dbl&gt;, FP.gt &lt;int&gt;, FP.al &lt;int&gt;, Subset.Size &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt; #   Subset.IS_CONF.Size &lt;dbl&gt;, Subset.Level &lt;dbl&gt;, TRUTH.TOTAL &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt; #   TRUTH.TOTAL.ti &lt;dbl&gt;, TRUTH.TOTAL.tv &lt;dbl&gt;, TRUTH.TOTAL.het &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt; #   TRUTH.TOTAL.homalt &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, TRUTH.TP &lt;int&gt;, TRUTH.TP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt; #   TRUTH.TP.tv &lt;dbl&gt;, TRUTH.TP.het &lt;dbl&gt;, TRUTH.TP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt; #   TRUTH.TP.TiTv_ratio &lt;dbl&gt;, TRUTH.TP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt; #   TRUTH.FN &lt;int&gt;, TRUTH.FN.ti &lt;dbl&gt;, TRUTH.FN.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co">#&gt; #   TRUTH.FN.het &lt;dbl&gt;, TRUTH.FN.homalt &lt;dbl&gt;, TRUTH.FN.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="co">#&gt; #   TRUTH.FN.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL &lt;int&gt;, QUERY.TOTAL.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"><span class="co">#&gt; #   QUERY.TOTAL.tv &lt;dbl&gt;, QUERY.TOTAL.het &lt;dbl&gt;, QUERY.TOTAL.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"><span class="co">#&gt; #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-25" data-line-number="25"><span class="co">#&gt; #   QUERY.TP &lt;int&gt;, QUERY.TP.ti &lt;dbl&gt;, QUERY.TP.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-26" data-line-number="26"><span class="co">#&gt; #   QUERY.TP.het &lt;dbl&gt;, QUERY.TP.homalt &lt;dbl&gt;, QUERY.TP.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"><span class="co">#&gt; #   QUERY.TP.het_hom_ratio &lt;dbl&gt;, QUERY.FP &lt;int&gt;, QUERY.FP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-28" data-line-number="28"><span class="co">#&gt; #   QUERY.FP.tv &lt;dbl&gt;, QUERY.FP.het &lt;dbl&gt;, QUERY.FP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-29" data-line-number="29"><span class="co">#&gt; #   QUERY.FP.TiTv_ratio &lt;dbl&gt;, QUERY.FP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"><span class="co">#&gt; #   QUERY.UNK &lt;int&gt;, QUERY.UNK.ti &lt;dbl&gt;, QUERY.UNK.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"><span class="co">#&gt; #   QUERY.UNK.het &lt;dbl&gt;, QUERY.UNK.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"><span class="co">#&gt; #   QUERY.UNK.TiTv_ratio &lt;dbl&gt;, QUERY.UNK.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="co">#&gt; #   file_source &lt;chr&gt;</span></a></code></pre></div>
</div>
<div id="aggregating-results-from-multiple-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregating-results-from-multiple-samples" class="anchor"></a>Aggregating results from multiple samples</h2>
<p>Often, we will want to inspect results from multiple samples, which we can achieve by passing a samplesheet to happyR:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># define happyr samplesheet</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">extdata_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">samplesheet &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span>(<span class="st">"group_id,replicate_id,happy_prefix</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="st">PCR-Free,NA12878-I30,NA12878-I30_S1</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="st">PCR-Free,NA12878-I33,NA12878-I33_S1</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="st">Nano,NA12878-R1,NA12878-R1_S1</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="st">Nano,NA12878-R2,NA12878-R2_S1</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="st">"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="kw">mutate</span>(<span class="dt">happy_prefix =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"%s/%s"</span>, extdata_dir, happy_prefix))</a>
<a class="sourceLine" id="cb5-10" data-line-number="10"></a>
<a class="sourceLine" id="cb5-11" data-line-number="11">samplesheet</a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt; # A tibble: 4 x 3</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt;   group_id replicate_id happy_prefix                                       </span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;        &lt;chr&gt;                                              </span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; 1 PCR-Free NA12878-I30  /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; 2 PCR-Free NA12878-I33  /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; 3 Nano     NA12878-R1   /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; 4 Nano     NA12878-R2   /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co"># load hap.py results</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21">hap_samplesheet &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_samplesheet_.html">read_samplesheet_</a></span>(samplesheet)</a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; Processing /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I30_S1</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; Reading summary table</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; Reading extended table</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="co">#&gt; Reading precision-recall curve data</span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I30_S1.roc.Locations.SNP.csv.gz</span></a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I30_S1.roc.Locations.SNP.PASS.csv.gz</span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I30_S1.roc.Locations.SNP.SEL.csv.gz</span></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="co">#&gt; Processing /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I33_S1</span></a>
<a class="sourceLine" id="cb5-30" data-line-number="30"><span class="co">#&gt; Reading summary table</span></a>
<a class="sourceLine" id="cb5-31" data-line-number="31"><span class="co">#&gt; Reading extended table</span></a>
<a class="sourceLine" id="cb5-32" data-line-number="32"><span class="co">#&gt; Reading precision-recall curve data</span></a>
<a class="sourceLine" id="cb5-33" data-line-number="33"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I33_S1.roc.Locations.SNP.csv.gz</span></a>
<a class="sourceLine" id="cb5-34" data-line-number="34"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I33_S1.roc.Locations.SNP.PASS.csv.gz</span></a>
<a class="sourceLine" id="cb5-35" data-line-number="35"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-I33_S1.roc.Locations.SNP.SEL.csv.gz</span></a>
<a class="sourceLine" id="cb5-36" data-line-number="36"><span class="co">#&gt; Processing /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R1_S1</span></a>
<a class="sourceLine" id="cb5-37" data-line-number="37"><span class="co">#&gt; Reading summary table</span></a>
<a class="sourceLine" id="cb5-38" data-line-number="38"><span class="co">#&gt; Reading extended table</span></a>
<a class="sourceLine" id="cb5-39" data-line-number="39"><span class="co">#&gt; Reading precision-recall curve data</span></a>
<a class="sourceLine" id="cb5-40" data-line-number="40"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R1_S1.roc.Locations.SNP.csv.gz</span></a>
<a class="sourceLine" id="cb5-41" data-line-number="41"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R1_S1.roc.Locations.SNP.PASS.csv.gz</span></a>
<a class="sourceLine" id="cb5-42" data-line-number="42"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R1_S1.roc.Locations.SNP.SEL.csv.gz</span></a>
<a class="sourceLine" id="cb5-43" data-line-number="43"><span class="co">#&gt; Processing /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R2_S1</span></a>
<a class="sourceLine" id="cb5-44" data-line-number="44"><span class="co">#&gt; Reading summary table</span></a>
<a class="sourceLine" id="cb5-45" data-line-number="45"><span class="co">#&gt; Reading extended table</span></a>
<a class="sourceLine" id="cb5-46" data-line-number="46"><span class="co">#&gt; Reading precision-recall curve data</span></a>
<a class="sourceLine" id="cb5-47" data-line-number="47"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R2_S1.roc.Locations.SNP.csv.gz</span></a>
<a class="sourceLine" id="cb5-48" data-line-number="48"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R2_S1.roc.Locations.SNP.PASS.csv.gz</span></a>
<a class="sourceLine" id="cb5-49" data-line-number="49"><span class="co">#&gt; Missing file: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/happyR/extdata/NA12878-R2_S1.roc.Locations.SNP.SEL.csv.gz</span></a>
<a class="sourceLine" id="cb5-50" data-line-number="50"><span class="co"># or directly from a samplesheet.csv</span></a>
<a class="sourceLine" id="cb5-51" data-line-number="51"><span class="co"># hap_samplesheet &lt;- read_samplesheet(samplesheet_path = "/path/to/happyr_samplesheet.csv")</span></a>
<a class="sourceLine" id="cb5-52" data-line-number="52"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(hap_samplesheet)</a>
<a class="sourceLine" id="cb5-53" data-line-number="53"><span class="co">#&gt; [1] "happy_samplesheet"</span></a>
<a class="sourceLine" id="cb5-54" data-line-number="54"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(hap_samplesheet)</a>
<a class="sourceLine" id="cb5-55" data-line-number="55"><span class="co">#&gt; [1] "samplesheet" "results"</span></a></code></pre></div>
<p><code>hap_samplesheet</code> is a <code>happy_samplesheet</code> object that contains the following fields:</p>
<ul>
<li>
<code>samplesheet</code>: a data.frame with the original samplesheet</li>
<li>
<code>results</code>: a <code>happy_result_list</code> that contains individual <code>happy_result</code> objects</li>
<li>
<code>ids</code>: a vector of result ids</li>
</ul>
<p>We can query <code>samplesheet</code> and <code>ids</code> fields by directly accessing the relevant list items:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">hap_samplesheet<span class="op">$</span>samplesheet <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; # A tibble: 4 x 3</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt;   group_id replicate_id happy_prefix                                       </span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;        &lt;chr&gt;                                              </span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; 1 PCR-Free NA12878-I30  /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; 2 PCR-Free NA12878-I33  /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; 3 Nano     NA12878-R1   /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; 4 Nano     NA12878-R2   /Library/Frameworks/R.framework/Versions/3.4/Resou…</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">hap_samplesheet<span class="op">$</span>ids <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; NULL</span></a></code></pre></div>
<p>And access aggregate hap.py results with the <code>extract_results</code> function:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extract_results.html">extract_results</a></span>(hap_samplesheet<span class="op">$</span>results, <span class="dt">table =</span> <span class="st">"summary"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">summary <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#&gt; # A tibble: 6 x 18</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt;   Type  Filter TRUTH.TOTAL TRUTH.TP TRUTH.FN QUERY.TOTAL QUERY.FP QUERY.UNK</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt; 1 INDEL ALL           8330     7655      675       15820      691      7319</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">#&gt; 2 INDEL PASS          8330     7547      783       13324      397      5221</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt; 3 SNP   ALL          48280    46396     1884       61177      342     14416</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt; 4 SNP   PASS         48280    45606     2674       52699      152      6921</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">#&gt; 5 INDEL ALL           8330     7629      701       15558      660      7119</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">#&gt; 6 INDEL PASS          8330     7546      784       13362      394      5270</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">#&gt; # … with 10 more variables: FP.gt &lt;dbl&gt;, METRIC.Recall &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">#&gt; #   METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;, METRIC.F1_Score &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">#&gt; #   TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">#&gt; #   happy_prefix &lt;chr&gt;</span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co"># see the extract_results documentation for a list of possible values for the table argument</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#set-up">Set up</a></li>
      <li><a href="#loading-and-querying-individual-hap-py-results">Loading and querying individual hap.py results</a></li>
      <li><a href="#aggregating-results-from-multiple-samples">Aggregating results from multiple samples</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
