<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Precision-recall curves • happyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">happyR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Precision-recall_curves.html">Precision-recall curves</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Precision-recall curves</h1>
                        <h4 class="author">Ben Moore</h4>
            
            <h4 class="date">2017-06-28</h4>
          </div>

    
    
<div class="contents">
<p>One useful set of outputs of that hap.py generates is precision and recall curve data (for information on how to generate this data, see the <a href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md#roc-curves">hap.py docs</a>). Once you’ve run hap.py, <code>happyR</code> contains a few helper functions to load, access and subset this precision-recall data from raw output files.</p>
<p>This vignette contains a few quick examples of loading data, extracting relevant subsets and plotting with <code>ggplot2</code>.</p>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>First load the example data that comes with the package, and load packages / set options for future chunks:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(happyR)

happy_input &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"happy_demo.summary.csv"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)
happy_prefix &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">".summary.csv"</span>, <span class="st">""</span>, happy_input)

hapdata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_happy.html">read_happy</a></span>(happy_prefix, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)

<span class="co"># other packages and options</span>
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(magrittr)
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme_get">theme_set</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>())</code></pre></div>
<div id="real-data" class="section level3">
<h3 class="hasAnchor">
<a href="#real-data" class="anchor"></a>Real data</h3>
<p>This vignette uses example data, in real applications you’ll want to load your own results. For example if you’d run hap.py with the output prefix <code>"happy_output/na12878"</code>:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">/opt/hap.py</span> -r <span class="ot">$reference</span> -f <span class="ot">$confident_regions</span> -o <span class="st">"happy_output/na12878"</span> <span class="ot">$truth</span> <span class="ot">$query</span></code></pre></div>
<p>Load this data into R with <code>happyR</code> as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">happy_data &lt;-<span class="st"> </span>happyR::<span class="kw"><a href="../reference/read_happy.html">read_happy</a></span>(<span class="st">"happy_output/na12878"</span>)</code></pre></div>
</div>
</div>
<div id="lazy-loading-large-files" class="section level2">
<h2 class="hasAnchor">
<a href="#lazy-loading-large-files" class="anchor"></a>Lazy-loading large files</h2>
<p>PR data associated with this hap.py result is lazy-loaded; that is, the data will only be read into R once it is accessed. This saves time that would otherwise be spent reading lots of data that’s never used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Large 'ALL' PR data isn't loaded yet and the results object isn't too big</span>
if (<span class="kw">require</span>(pryr))
  pryr::<span class="kw"><a href="http://www.rdocumentation.org/packages/pryr/topics/object_size">object_size</a></span>(hapdata)
<span class="co">#&gt; 21.4 MB</span></code></pre></div>
<p>Any access, even just a <code>head</code>, evaluates the promise and loads the remaining data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># view part of all PR data, evaluating the promise</span>
knitr::<span class="kw">kable</span>(<span class="kw">head</span>(hapdata$pr_curve$all[,<span class="dv">1</span>:<span class="dv">9</span>]))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Type</th>
<th align="left">Subtype</th>
<th align="left">Subset</th>
<th align="left">Filter</th>
<th align="left">Genotype</th>
<th align="left">QQ.Field</th>
<th align="left">QQ</th>
<th align="right">METRIC.Recall</th>
<th align="right">METRIC.Precision</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">*</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">0.000000</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">10.900000</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">1000.669983</td>
<td align="right">0.497818</td>
<td align="right">0.961003</td>
</tr>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">1001.200012</td>
<td align="right">0.497706</td>
<td align="right">0.960987</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="left">10015.700195</td>
<td align="right">0.000895</td>
<td align="right">1.000000</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The hapdata object now includes the full PR data so is a bit bigger</span>
if (<span class="kw">require</span>(pryr))
  pryr::<span class="kw"><a href="http://www.rdocumentation.org/packages/pryr/topics/object_size">object_size</a></span>(hapdata)
<span class="co">#&gt; 85 MB</span></code></pre></div>
<p>If you don’t find this useful, turn it off by passing <code>lazy = FALSE</code> to <code>read_happy</code>.</p>
</div>
<div id="subset-pr-curves" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-pr-curves" class="anchor"></a>Subset PR curves</h2>
<p>Subsetting the correct PR curve data can be a bit unwieldy:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset for short insertions 1 - 5 bp in length</span>
short_ins1 &lt;-<span class="st"> </span><span class="kw">subset</span>(hapdata$pr_curve$all, Filter ==<span class="st"> "ALL"</span> &amp;<span class="st"> </span>Subset ==<span class="st"> "*"</span> &amp;<span class="st"> </span>Subtype ==<span class="st"> "I1_5"</span>)</code></pre></div>
<p>happyR offers <code>pr_data</code> to make this a bit simpler:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">short_ins2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"indel"</span>, <span class="dt">subtype =</span> <span class="st">"I1_5"</span>)

<span class="co"># check they give the same results</span>
<span class="kw">all.equal</span>(short_ins1, short_ins2)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<div id="deletions-by-length" class="section level3">
<h3 class="hasAnchor">
<a href="#deletions-by-length" class="anchor"></a>Deletions by length</h3>
<p>Insertion subtypes are of the form: <code>[IDC]length_range</code> where the first letter indicates the variant classification: <code>I</code> insertion; <code>D</code> deletion; and <code>C</code> complex. Hap.py bins the lengths of these records into ranges by allele length in basepairs: <code>1_5</code>, <code>6_15</code> and <code>16_PLUS</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get only deletions for all length ranges</span>
del_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"indel"</span>, 
                  <span class="dt">subtype =</span> <span class="kw">c</span>(<span class="st">"D1_5"</span>, <span class="st">"D6_15"</span>, <span class="st">"D16_PLUS"</span>))

<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(del_pr, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Subtype)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_cartesian">coord_cartesian</a></span>(<span class="dt">ylim =</span> <span class="kw">c</span>(.<span class="dv">75</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Longer deletions are more difficult to call accurately"</span>)</code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/del_length-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="snvs-at-truthset-boundaries" class="section level3">
<h3 class="hasAnchor">
<a href="#snvs-at-truthset-boundaries" class="anchor"></a>SNVs at truthset boundaries</h3>
<p>Hap.py v0.3.7+ writes subsets <code>TS_contained</code> and <code>TS_boundary</code> by default. These contain truth variants which are either comfortably contained or near the boundary of confident regions, respectively. In some truthsets, those in <code>TS_boundary</code> will show worse performance metrics due to issues with variant representation or a partial haplotype description in the truthset. Alternatively, <code>TS_contained</code> can be an easier-to-call subset of truth variants where the surrounding region was able to be fully-characterised.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snv_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"snv"</span>,
                  <span class="dt">subset =</span> <span class="kw">c</span>(<span class="st">"TS_contained"</span>, <span class="st">"TS_boundary"</span>))

<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(snv_pr, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Subset)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_cartesian">coord_cartesian</a></span>(<span class="dt">ylim =</span> <span class="kw">c</span>(.<span class="dv">8</span>, <span class="dv">1</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"TS_contained performance is higher than TS_boundary"</span>)</code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/snv_boundary-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#example-data">Example data</a></li>
      <li><a href="#lazy-loading-large-files">Lazy-loading large files</a></li>
      <li><a href="#subset-pr-curves">Subset PR curves</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
