<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Precision-recall curves • happyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Precision-recall curves">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">happyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Precision-recall_curves.html">Precision-recall curves</a>
    </li>
    <li>
      <a href="../articles/loading_data.html">Loading hap.py results</a>
    </li>
    <li>
      <a href="../articles/stratified_counts.html">Analysing stratified counts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Precision-recall curves</h1>
                        <h4 class="author">Ben Moore</h4>
            
            <h4 class="date">2019-07-09</h4>
      
      
      <div class="hidden name"><code>Precision-recall_curves.Rmd</code></div>

    </div>

    
    
<p>One useful set of outputs of that hap.py generates is precision and recall curve data (for information on how to generate this data, see the <a href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md#roc-curves">hap.py docs</a>). Once you’ve run hap.py, <code>happyR</code> contains a few helper functions to load, access and subset this precision-recall data from raw output files.</p>
<p>This vignette contains a few quick examples of loading data, extracting relevant subsets and plotting with <code>ggplot2</code>.</p>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>First load the example data that comes with the package, and load packages / set options for future chunks:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(happyR)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">happy_input &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"happy_demo.summary.csv"</span>, <span class="dt">package =</span> <span class="st">"happyR"</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">happy_prefix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">sub</a></span>(<span class="st">".summary.csv"</span>, <span class="st">""</span>, happy_input)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">hapdata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_happy.html">read_happy</a></span>(happy_prefix, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co"># other packages and options</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>())</a></code></pre></div>
<div id="real-data" class="section level3">
<h3 class="hasAnchor">
<a href="#real-data" class="anchor"></a>Real data</h3>
<p>This vignette uses example data, in real applications you’ll want to load your own results. For example if you’d run hap.py with the output prefix <code>"happy_output/na12878"</code>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">/opt/hap.py</span> -r <span class="va">$reference</span> -f <span class="va">$confident_regions</span> -o <span class="st">"happy_output/na12878"</span> <span class="va">$truth</span> <span class="va">$query</span></a></code></pre></div>
<p>Load this data into R with <code>happyR</code> as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">happy_data &lt;-<span class="st"> </span>happyR<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/happyR/topics/read_happy">read_happy</a></span>(<span class="st">"happy_output/na12878"</span>)</a></code></pre></div>
</div>
</div>
<div id="lazy-loading-large-files" class="section level2">
<h2 class="hasAnchor">
<a href="#lazy-loading-large-files" class="anchor"></a>Lazy-loading large files</h2>
<p>PR data associated with this hap.py result is lazy-loaded; that is, the data will only be read into R once it is accessed. This saves time that would otherwise be spent reading lots of data that’s never used.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Large 'ALL' PR data isn't loaded yet and the results object isn't too big</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(pryr))</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">  pryr<span class="op">::</span><span class="kw">object_size</span>(hapdata)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt; Warning in library(package, lib.loc = lib.loc, character.only = TRUE,</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; logical.return = TRUE, : there is no package called 'pryr'</span></a></code></pre></div>
<p>Any access, even just a <code>head</code>, evaluates the promise and loads the remaining data:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># view part of all PR data, evaluating the promise</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(hapdata<span class="op">$</span>pr_curve<span class="op">$</span>all[,<span class="dv">1</span><span class="op">:</span><span class="dv">9</span>]))</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Type</th>
<th align="left">Subtype</th>
<th align="left">Subset</th>
<th align="left">Filter</th>
<th align="left">Genotype</th>
<th align="left">QQ.Field</th>
<th align="right">QQ</th>
<th align="right">METRIC.Recall</th>
<th align="right">METRIC.Precision</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">NA</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">0.00</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">10.90</td>
<td align="right">0.877140</td>
<td align="right">0.958635</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">1000.67</td>
<td align="right">0.497818</td>
<td align="right">0.961003</td>
</tr>
<tr class="odd">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">1001.20</td>
<td align="right">0.497706</td>
<td align="right">0.960987</td>
</tr>
<tr class="even">
<td align="left">INDEL</td>
<td align="left">*</td>
<td align="left">*</td>
<td align="left">ALL</td>
<td align="left">*</td>
<td align="left">QUAL</td>
<td align="right">10015.70</td>
<td align="right">0.000895</td>
<td align="right">1.000000</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># The hapdata object now includes the full PR data so is a bit bigger</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(pryr))</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">  pryr<span class="op">::</span><span class="kw">object_size</span>(hapdata)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; Loading required package: pryr</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; Warning in library(package, lib.loc = lib.loc, character.only = TRUE,</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; logical.return = TRUE, : there is no package called 'pryr'</span></a></code></pre></div>
<p>If you don’t find this useful, turn it off by passing <code>lazy = FALSE</code> to <code>read_happy</code>.</p>
</div>
<div id="high-level-pr-curves" class="section level2">
<h2 class="hasAnchor">
<a href="#high-level-pr-curves" class="anchor"></a>High-level PR curves</h2>
<p>PR curves show how precision and recall vary with a changing threshold, in this example it’s over a range of quality score thresholds applied to a set of variants: as the threshold increases, the remaining variant set is less comprehensive (lower recall) but typically contains fewer false positives (higher precision).</p>
<p>This gets more complicated when <code>PASS</code> records aren’t set solely by a single quality score threshold, but by multiple independent filters (e.g. high-depth, genomic context, etc.). These interactions can’t be fully captured by drawing a PR curve only on a quality threshold:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># using happyR::pr_data to simplify subsetting:</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">all_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co"># this gets PR curve starting at ALL point, equivalent to base:</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co"># all_pr &lt;- subset(hap_result$pr_curve$all, Filter == "ALL" &amp; Subtype == "*" &amp; Subset == "*")</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(all_pr, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Type)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">data =</span> hapdata<span class="op">$</span>summary) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">6</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">95</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"ALL PR curve might not hit the PASS point"</span>)</a></code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/all_pr-1.png" width="700"></p>
<p>Hap.py accounts for this by generating a selectively-filtered PR curve (<code>SEL</code>). First these independent filters are applied to the variant set, then we can draw the remaining PR curve using the quality score threshold. This gives a more accurate view of how changing the quality score threshold can impact precision and recall.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># selectively filtered PR curve</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">filter =</span> <span class="st">"SEL"</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co"># link this to the ALL point</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5">pr &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(pr, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/subset">subset</a></span>(hapdata<span class="op">$</span>summary, Filter <span class="op">==</span><span class="st"> "ALL"</span>))</a>
<a class="sourceLine" id="cb8-6" data-line-number="6"></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pr, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Type)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">data =</span> hapdata<span class="op">$</span>summary) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">6</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">95</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Selectively-filtered PR curve reliably hits PASS"</span>)</a></code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/sel_pr-1.png" width="700"></p>
</div>
<div id="subset-pr-curves" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-pr-curves" class="anchor"></a>Subset PR curves</h2>
<p>Subsetting the correct PR curve data can be a bit unwieldy:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># subset for short insertions 1 - 5 bp in length</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">short_ins1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/subset">subset</a></span>(hapdata<span class="op">$</span>pr_curve<span class="op">$</span>all, Filter <span class="op">==</span><span class="st"> "ALL"</span> <span class="op">&amp;</span><span class="st"> </span>Subset <span class="op">==</span><span class="st"> "*"</span> <span class="op">&amp;</span><span class="st"> </span>Subtype <span class="op">==</span><span class="st"> "I1_5"</span>)</a></code></pre></div>
<p>happyR offers <code>pr_data</code> to make this a bit simpler:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">short_ins2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"indel"</span>, <span class="dt">subtype =</span> <span class="st">"I1_5"</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co"># check they give the same results</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all.equal">all.equal</a></span>(short_ins1, short_ins2)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co">#&gt; [1] TRUE</span></a></code></pre></div>
<div id="deletions-by-length" class="section level3">
<h3 class="hasAnchor">
<a href="#deletions-by-length" class="anchor"></a>Deletions by length</h3>
<p>Insertion subtypes are of the form: <code>[IDC]length_range</code> where the first letter indicates the variant classification: <code>I</code> insertion; <code>D</code> deletion; and <code>C</code> complex. Hap.py bins the lengths of these records into ranges by allele length in basepairs: <code>1_5</code>, <code>6_15</code> and <code>16_PLUS</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># get only deletions for all length ranges</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">del_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"indel"</span>, </a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                  <span class="dt">subtype =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"D1_5"</span>, <span class="st">"D6_15"</span>, <span class="st">"D16_PLUS"</span>))</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(del_pr, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Subtype)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">ylim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">75</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Longer deletions are more difficult to call accurately"</span>)</a></code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/del_length-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="snvs-at-truthset-boundaries" class="section level3">
<h3 class="hasAnchor">
<a href="#snvs-at-truthset-boundaries" class="anchor"></a>SNVs at truthset boundaries</h3>
<p>Hap.py v0.3.7+ writes subsets <code>TS_contained</code> and <code>TS_boundary</code> by default. These contain truth variants which are either comfortably contained or near the boundary of confident regions, respectively. In some truthsets, those in <code>TS_boundary</code> will show worse performance metrics due to issues with variant representation or a partial haplotype description in the truthset. Alternatively, <code>TS_contained</code> can be an easier-to-call subset of truth variants where the surrounding region was able to be fully-characterised.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">snv_pr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pr_data.html">pr_data</a></span>(hapdata, <span class="dt">var_type =</span> <span class="st">"snv"</span>,</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">                  <span class="dt">subset =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"TS_contained"</span>, <span class="st">"TS_boundary"</span>))</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(snv_pr, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> METRIC.Recall, <span class="dt">y =</span> METRIC.Precision, <span class="dt">col =</span> Subset)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">ylim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">8</span>, <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"TS_contained performance is higher than TS_boundary"</span>)</a></code></pre></div>
<p><img src="Precision-recall_curves_files/figure-html/snv_boundary-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#example-data">Example data</a></li>
      <li><a href="#lazy-loading-large-files">Lazy-loading large files</a></li>
      <li><a href="#high-level-pr-curves">High-level PR curves</a></li>
      <li><a href="#subset-pr-curves">Subset PR curves</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
