<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysing stratified counts • happyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysing stratified counts">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">happyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/loading_data.html">Loading hap.py results</a>
    </li>
    <li>
      <a href="../articles/precision-recall_curves.html">Precision-recall curves</a>
    </li>
    <li>
      <a href="../articles/stratified_counts.html">Analysing stratified counts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analysing stratified counts</h1>
                        <h4 class="author">Mar Gonzalez-Porta</h4>
            
            <h4 class="date">2019-06-09</h4>
      
      
      <div class="hidden name"><code>stratified_counts.Rmd</code></div>

    </div>

    
    
<p>In the present vignette we will be comparing stratified recall between a set of PCR-Free vs. Nano builds from NA12878. In addition, we will discuss our approach to calibrate our HDI estimates using a simulated dataset.</p>
<p>At this point, we might want to compare calling performance across groups (PCR-Free vs. Nano) and/or across genomic subsets (e.g. XX vs. XX). However, replicate variability and variable subset sizes pose a challenge when interpreting the observed results. E.g. if we observe the counts from the table below across 4 different replicates, can we confidently claim that there is a difference in performance across conditions for subset A? And can we say that there is a difference across subsets for condition 1?</p>
<table class="table">
<thead><tr class="header">
<th>SUBSET</th>
<th>TP.CONDITION_1</th>
<th>TP.CONDITION_2</th>
<th>TRUTH.TOTAL</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>81,82,82,82</td>
<td>74,75,76,75</td>
<td>82</td>
</tr>
<tr class="even">
<td>B</td>
<td>2,2,2,2</td>
<td>2,2,2,2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>Here we propose a method to compute highest-posterior density intervals (HDIs) for performance metrics that simultaneously take into account subset size and replicate variability, and eventually answer such questions. The method is based on the following steps:</p>
<ol style="list-style-type: decimal">
<li>Model counts using Negative Binomial sampling</li>
</ol>
<p><span class="math display">\[
k_{ij} \sim NB(\rho_{i}*n_{ij}, \sigma)
\]</span> where</p>
<p><span class="math inline">\(k_{ij}\)</span> = observed successes for subset i in replicate j; <span class="math inline">\(\rho_{i}\)</span> = success rate for subset i; <span class="math inline">\(n_{ij}\)</span> = total counts for subset i in replicate j; <span class="math inline">\(\sigma\)</span> = dispersion parameter; <span class="math inline">\(\rho_{i} \sim Beta(1, 1)\)</span> = <span class="math inline">\(\rho\)</span> prior; <span class="math inline">\(\sigma \sim Normal(\mu, \sigma^2)\)</span> = <span class="math inline">\(\sigma\)</span> prior</p>
<ol start="2" style="list-style-type: decimal">
<li>Use MCMC to obtain posterior distributions for the success rate, per replicate</li>
<li>Compute highest density intervals (HDI) by sampling from the aggregate posterior</li>
</ol>
<div id="set-up" class="section level2">
<h2 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set up</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(happyR)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">theme_set</span>(<span class="kw">theme_minimal</span>())</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/source">source</a></span>(<span class="st">"stratified_counts_util.R"</span>)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">42</span>)</a></code></pre></div>
</div>
<div id="comparing-stratified-counts-across-analysis-groups" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-stratified-counts-across-analysis-groups" class="anchor"></a>Comparing stratified counts across analysis groups</h2>
<p>For our analyses, we will be relying on pre-processed data that we have generated by running hap.py on VCFs from <a href="https://basespace.illumina.com/datacentral">BaseSpace Data Central</a>. 5 pcrfrer, 8 nano - wgsv8.1 + grch38</p>
<div id="obtaining-stratified-counts" class="section level3">
<h3 class="hasAnchor">
<a href="#obtaining-stratified-counts" class="anchor"></a>Obtaining stratified counts</h3>
<p>With hap.py, it is possible to evaluate calling performance across various genomic subsets</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">hap.py</span> <span class="va">${TRUTH_VCF}</span> <span class="va">${QUERY_VCF}</span> -o <span class="va">${OUTPUT_PREFIX}</span> -f <span class="va">${CONFIDENT_REGIONS}</span> \</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">    --threads 40 --write-counts --reference <span class="va">${REF}</span> \</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">    --stratification <span class="va">${STRATIFICATION_CONFIG}</span> \</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">    --roc QUAL --roc-filter LowQual --no-json</a></code></pre></div>
<p>TODO: explain stratification regions used</p>
<p>Note that for speed reasons, we have focused our analyses on <code>chr22</code>, and have subsetted the truth VCF and stratification BEDs accordingly.</p>
</div>
<div id="loading-demo-results-into-r" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-demo-results-into-r" class="anchor"></a>Loading demo results into R</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">data_dir &lt;-<span class="st"> "demo_data.pcrfree_vs_nano/filtered_happy/"</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">samplesheet &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span>(<span class="st">"group_id,replicate_id,happy_prefix</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="st">PCR-Free,NA12878-I30,NA12878-I30_S1</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="st">PCR-Free,NA12878-I33,NA12878-I33_S1</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="st">PCR-Free,NA12878-I47,NA12878-I47_S1</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="st">PCR-Free,NA12878-I67,NA12878-I67_S1</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="st">PCR-Free,NA12878-I82,NA12878-I82_S1</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="st">PCR-Free,NA12878-I85,NA12878-I85_S1</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="st">Nano,NA12878-R1,NA12878-R1_S1</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="st">Nano,NA12878-R2,NA12878-R2_S1</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="st">Nano,NA12878-R3,NA12878-R3_S1</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="st">Nano,NA12878-R4,NA12878-R4_S1</span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="st">Nano,NA12878-R5,NA12878-R5_S1</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="st">Nano,NA12878-R6,NA12878-R6_S1</span></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="st">Nano,NA12878-R7,NA12878-R7_S1</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="st">Nano,NA12878-R8,NA12878-R8_S1</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="st">"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="kw">mutate</span>(<span class="dt">happy_prefix =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"%s/%s"</span>, data_dir, happy_prefix))</a>
<a class="sourceLine" id="cb3-19" data-line-number="19"></a>
<a class="sourceLine" id="cb3-20" data-line-number="20">hap_samplesheet &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_samplesheet_.html">read_samplesheet_</a></span>(samplesheet)</a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">saveRDS</a></span>(hap_samplesheet, <span class="dt">file =</span> <span class="st">"demo_data.Rds"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># load the demo dataset from a pre-saved Rds</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">hap_samplesheet &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="st">"demo_data.Rds"</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(hap_samplesheet)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt; [1] "happy_samplesheet"</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># load pre-computed HDI</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">demo_hdi &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">"demo_hdi.csv"</span>)</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; Parsed with column specification:</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; cols(</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt;   i = col_double(),</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt;   subset = col_character(),</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt;   n.obs = col_double(),</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt;   prior.mean = col_double(),</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt;   obs.min = col_double(),</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt;   obs.max = col_double(),</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt;   posterior.mean = col_double(),</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt;   posterior.hdi.low = col_double(),</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt;   posterior.hdi.high = col_double(),</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt;   .type_group = col_character()</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt; )</span></a></code></pre></div>
<p>Once we have a <code>happy_samplesheet</code> object we can rely on the <code>extract_metrics</code> function to access our metrics of interest. Since hap.py saves stratified counts under <code>*extended.csv</code> files, we can use the following:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">stratified_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extract_results.html">extract_results</a></span>(hap_samplesheet<span class="op">$</span>results, <span class="dt">table =</span> <span class="st">"extended"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="st">  </span><span class="co"># focus on PASS calls in level 0 subsets</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Subtype <span class="op">==</span><span class="st"> "*"</span>, Filter <span class="op">==</span><span class="st"> "PASS"</span>, Subset.Level <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="dt">pattern =</span> <span class="st">"TS*"</span>, Subset)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="st">  </span><span class="kw">inner_join</span>(hap_samplesheet<span class="op">$</span>samplesheet) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.type_group =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(Type, group_id))</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; Joining, by = "happy_prefix"</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"></a>
<a class="sourceLine" id="cb5-8" data-line-number="8">stratified_counts <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; # A tibble: 6 x 69</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt;   Type  Subtype Subset Filter Genotype QQ.Field    QQ METRIC.Recall</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;    &lt;dbl&gt;         &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt; 1 INDEL *       *      PASS   *        QUAL        NA         0.906</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt; 2 INDEL *       codin… PASS   *        QUAL        NA         1    </span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt; 3 INDEL *       high.… PASS   *        QUAL        NA         0.875</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; 4 INDEL *       high.… PASS   *        QUAL        NA         0.905</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; 5 INDEL *       repea… PASS   *        QUAL        NA         0.892</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; 6 SNP   *       *      PASS   *        QUAL        NA         0.945</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; # … with 61 more variables: METRIC.Precision &lt;dbl&gt;, METRIC.Frac_NA &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; #   METRIC.F1_Score &lt;dbl&gt;, FP.gt &lt;int&gt;, FP.al &lt;int&gt;, Subset.Size &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; #   Subset.IS_CONF.Size &lt;dbl&gt;, Subset.Level &lt;dbl&gt;, TRUTH.TOTAL &lt;int&gt;,</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; #   TRUTH.TOTAL.ti &lt;dbl&gt;, TRUTH.TOTAL.tv &lt;dbl&gt;, TRUTH.TOTAL.het &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; #   TRUTH.TOTAL.homalt &lt;dbl&gt;, TRUTH.TOTAL.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; #   TRUTH.TOTAL.het_hom_ratio &lt;dbl&gt;, TRUTH.TP &lt;int&gt;, TRUTH.TP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; #   TRUTH.TP.tv &lt;dbl&gt;, TRUTH.TP.het &lt;dbl&gt;, TRUTH.TP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="co">#&gt; #   TRUTH.TP.TiTv_ratio &lt;dbl&gt;, TRUTH.TP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="co">#&gt; #   TRUTH.FN &lt;int&gt;, TRUTH.FN.ti &lt;dbl&gt;, TRUTH.FN.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; #   TRUTH.FN.het &lt;dbl&gt;, TRUTH.FN.homalt &lt;dbl&gt;, TRUTH.FN.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"><span class="co">#&gt; #   TRUTH.FN.het_hom_ratio &lt;dbl&gt;, QUERY.TOTAL &lt;int&gt;, QUERY.TOTAL.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="co">#&gt; #   QUERY.TOTAL.tv &lt;dbl&gt;, QUERY.TOTAL.het &lt;dbl&gt;, QUERY.TOTAL.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-30" data-line-number="30"><span class="co">#&gt; #   QUERY.TOTAL.TiTv_ratio &lt;dbl&gt;, QUERY.TOTAL.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-31" data-line-number="31"><span class="co">#&gt; #   QUERY.TP &lt;int&gt;, QUERY.TP.ti &lt;dbl&gt;, QUERY.TP.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-32" data-line-number="32"><span class="co">#&gt; #   QUERY.TP.het &lt;dbl&gt;, QUERY.TP.homalt &lt;dbl&gt;, QUERY.TP.TiTv_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-33" data-line-number="33"><span class="co">#&gt; #   QUERY.TP.het_hom_ratio &lt;dbl&gt;, QUERY.FP &lt;int&gt;, QUERY.FP.ti &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-34" data-line-number="34"><span class="co">#&gt; #   QUERY.FP.tv &lt;dbl&gt;, QUERY.FP.het &lt;dbl&gt;, QUERY.FP.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-35" data-line-number="35"><span class="co">#&gt; #   QUERY.FP.TiTv_ratio &lt;dbl&gt;, QUERY.FP.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-36" data-line-number="36"><span class="co">#&gt; #   QUERY.UNK &lt;int&gt;, QUERY.UNK.ti &lt;dbl&gt;, QUERY.UNK.tv &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-37" data-line-number="37"><span class="co">#&gt; #   QUERY.UNK.het &lt;dbl&gt;, QUERY.UNK.homalt &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-38" data-line-number="38"><span class="co">#&gt; #   QUERY.UNK.TiTv_ratio &lt;dbl&gt;, QUERY.UNK.het_hom_ratio &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb5-39" data-line-number="39"><span class="co">#&gt; #   happy_prefix &lt;chr&gt;, group_id &lt;chr&gt;, replicate_id &lt;chr&gt;,</span></a>
<a class="sourceLine" id="cb5-40" data-line-number="40"><span class="co">#&gt; #   .type_group &lt;chr&gt;</span></a>
<a class="sourceLine" id="cb5-41" data-line-number="41"></a>
<a class="sourceLine" id="cb5-42" data-line-number="42"><span class="co"># stratified_counts %&gt;% </span></a>
<a class="sourceLine" id="cb5-43" data-line-number="43"><span class="co">#   select(Subset, Type, Subset.Size, TRUTH.TOTAL) %&gt;% </span></a>
<a class="sourceLine" id="cb5-44" data-line-number="44"><span class="co">#   unique() %&gt;% </span></a>
<a class="sourceLine" id="cb5-45" data-line-number="45"><span class="co">#   spread(key = Type, value = TRUTH.TOTAL) %&gt;% </span></a>
<a class="sourceLine" id="cb5-46" data-line-number="46"><span class="co">#   rename(TRUTH.TOTAL.INDEL = INDEL) %&gt;% </span></a>
<a class="sourceLine" id="cb5-47" data-line-number="47"><span class="co">#   rename(TRUTH.TOTAL.SNP = SNP)</span></a>
<a class="sourceLine" id="cb5-48" data-line-number="48"></a>
<a class="sourceLine" id="cb5-49" data-line-number="49">stratified_counts <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-50" data-line-number="50"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Subset <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"coding.exons.acmg.ensembl87"</span>, <span class="st">"repeat.masker"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-51" data-line-number="51"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(replicate_id <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"NA12878-R1"</span>, <span class="st">"NA12878-I30"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-52" data-line-number="52"><span class="st">  </span><span class="kw">select</span>(.type_group, Subset, replicate_id, TRUTH.TOTAL, TRUTH.TP, METRIC.Recall) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-53" data-line-number="53"><span class="st">  </span><span class="kw">arrange</span>(Subset, .type_group, replicate_id)</a>
<a class="sourceLine" id="cb5-54" data-line-number="54"><span class="co">#&gt; # A tibble: 8 x 6</span></a>
<a class="sourceLine" id="cb5-55" data-line-number="55"><span class="co">#&gt;   .type_group  Subset       replicate_id TRUTH.TOTAL TRUTH.TP METRIC.Recall</span></a>
<a class="sourceLine" id="cb5-56" data-line-number="56"><span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;        &lt;chr&gt;              &lt;int&gt;    &lt;int&gt;         &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-57" data-line-number="57"><span class="co">#&gt; 1 INDEL Nano   coding.exon… NA12878-R1             1        1         1    </span></a>
<a class="sourceLine" id="cb5-58" data-line-number="58"><span class="co">#&gt; 2 INDEL PCR-F… coding.exon… NA12878-I30            1        1         1    </span></a>
<a class="sourceLine" id="cb5-59" data-line-number="59"><span class="co">#&gt; 3 SNP Nano     coding.exon… NA12878-R1             6        6         1    </span></a>
<a class="sourceLine" id="cb5-60" data-line-number="60"><span class="co">#&gt; 4 SNP PCR-Free coding.exon… NA12878-I30            6        6         1    </span></a>
<a class="sourceLine" id="cb5-61" data-line-number="61"><span class="co">#&gt; 5 INDEL Nano   repeat.mask… NA12878-R1          5589     4447         0.796</span></a>
<a class="sourceLine" id="cb5-62" data-line-number="62"><span class="co">#&gt; 6 INDEL PCR-F… repeat.mask… NA12878-I30         5589     4984         0.892</span></a>
<a class="sourceLine" id="cb5-63" data-line-number="63"><span class="co">#&gt; 7 SNP Nano     repeat.mask… NA12878-R1         25224    23709         0.940</span></a>
<a class="sourceLine" id="cb5-64" data-line-number="64"><span class="co">#&gt; 8 SNP PCR-Free repeat.mask… NA12878-I30        25224    23681         0.939</span></a></code></pre></div>
<p>Comment on difference in subset sizes and truth.total.</p>
</div>
<div id="calculating-hdis" class="section level3">
<h3 class="hasAnchor">
<a href="#calculating-hdis" class="anchor"></a>Calculating HDIs</h3>
<p>TODO: dim_hdi.Rds</p>
<p>TODO: Also, a pre-computed dataset</p>
<p>From here, we can easily estimate HDIs for each performance metric using the helper functions attached to this vignette. Let’s use recall as an example:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">groups &lt;-<span class="st"> </span>stratified_counts <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(.type_group) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">demo_hdi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_along</a></span>(groups), <span class="cf">function</span>(i) {</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">  sel_group_type &lt;-<span class="st"> </span>groups[i]</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">  stan_result &lt;-<span class="st"> </span><span class="kw">sample_posterior</span>(<span class="dt">m =</span> stratified_counts <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(.type_group <span class="op">==</span><span class="st"> </span>sel_group_type),</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">                                  <span class="dt">successes_field =</span> <span class="st">"TRUTH.TP"</span>, <span class="dt">totals_field =</span> <span class="st">"TRUTH.TOTAL"</span>)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">  demo_hdi &lt;-<span class="st"> </span><span class="kw">estimate_hdi</span>(<span class="dt">r =</span> stan_result) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">.type_group =</span> sel_group_type)</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">  demo_hdi</a>
<a class="sourceLine" id="cb6-9" data-line-number="9">}) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="st">  </span><span class="kw">bind_rows</span>()</a>
<a class="sourceLine" id="cb6-11" data-line-number="11"></a>
<a class="sourceLine" id="cb6-12" data-line-number="12">demo_hdi <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="kw">write_csv</span>(demo_hdi, <span class="dt">path =</span> <span class="st">"demo_hdi.csv"</span>)</a></code></pre></div>
</div>
<div id="demo-results" class="section level3">
<h3 class="hasAnchor">
<a href="#demo-results" class="anchor"></a>Demo results</h3>
<p>And visualise the results using <code>ggplot2</code>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">dodge_width &lt;-<span class="st"> </span><span class="fl">0.7</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">demo_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> subset, <span class="dt">color =</span> .type_group)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="st">    </span><span class="co"># observed</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> obs.min, <span class="dt">ymax =</span> obs.max), </a>
<a class="sourceLine" id="cb7-6" data-line-number="6">                  <span class="dt">alpha =</span> <span class="fl">0.2</span>, <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width), <span class="dt">width =</span> <span class="dv">0</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="st">    </span><span class="co"># estimated</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> posterior.hdi.low, <span class="dt">ymax =</span> posterior.hdi.high), </a>
<a class="sourceLine" id="cb7-9" data-line-number="9">                  <span class="dt">size =</span> <span class="fl">0.5</span>, <span class="dt">alpha =</span> <span class="dv">1</span>, <span class="dt">width =</span> <span class="fl">0.6</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">y =</span> posterior.mean), <span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">shape =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb7-11" data-line-number="11">               <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">"METRIC.Recall"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">ylim</a></span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="st">    </span><span class="kw">coord_flip</span>()</a></code></pre></div>
<p><img src="stratified_counts_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
<div id="calibrating-hdis-with-simulated-data" class="section level2">
<h2 class="hasAnchor">
<a href="#calibrating-hdis-with-simulated-data" class="anchor"></a>Calibrating HDIs with simulated data</h2>
<p>One way to evaluate the accuracy of our HDIs is through simulated data, where we know what the true success rate is, and we can measure how often that value is captured by our HDI estimates. However, we do not want to rely on a simulation approach that is based solely on the model that we have used for estimation, since it might not be fully representative of real data.</p>
<div id="simulation-strategy" class="section level3">
<h3 class="hasAnchor">
<a href="#simulation-strategy" class="anchor"></a>Simulation strategy</h3>
<p>Let us take recall measurements across <span class="math inline">\(r\)</span> replicates as an example. For a subset with <span class="math inline">\(n\)</span> total variants we can classify each variant into one of these 3 groups:</p>
<ol style="list-style-type: decimal">
<li>the variant is consistently called across all of the replicates (always seen - A)</li>
<li>the variant is only called in a subset of the replicates (sometimes seen - S)</li>
<li>the variant is never called (never seen - N)</li>
</ol>
<p>We can then define our true recall <span class="math inline">\(\rho\)</span> as follows:</p>
<p><span class="math inline">\(\rho = (n*q_A*x_A + n*q_S*x_S + n*q_N*x_N) / n\)</span></p>
<p>where</p>
<p><span class="math inline">\(n\)</span>: total number of variants in the subset; <span class="math inline">\(q_A\)</span>: probability of having variants in group A; <span class="math inline">\(x_A\)</span>: recall in group A (will always be 1); <span class="math inline">\(q_S\)</span>: probability of having variants in group S; <span class="math inline">\(x_S\)</span>: recall in group S (variable); <span class="math inline">\(q_N\)</span>: probability of having variants in group N; <span class="math inline">\(x_N\)</span>: recall in group N (will always be 0).</p>
<p>From these parameters, we can simulate true positive counts in one sample by drawing from a multinomial distribution that attributes each of the <span class="math inline">\(n\)</span> variants into one of the groups using the defined probabilities <span class="math inline">\(\{q_A, q_S, q_N\}\)</span>. We can next simulate replicates from this sample by performing random binomial draws from each group with recalls <span class="math inline">\(\{x_A, x_S, x_N\}\)</span>, which in practice will be <span class="math inline">\(\{1, x_S, 0\}\)</span>. Note that the number of true positives for groups A and N will be fixed across replicates by definition, but we can still use <span class="math inline">\(x_S\)</span> to simulate replicate variability. As a result of this variability, our observed recall will differ slightly from the theoretical expectation <span class="math inline">\(\rho\)</span>.</p>
</div>
<div id="generating-the-simulated-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#generating-the-simulated-dataset" class="anchor"></a>Generating the simulated dataset</h3>
<p>In this vignette we will be working with pre-computed datasets:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">## simulated data</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">sim_data &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="dt">file =</span> <span class="st">"sim_data.csv"</span>, <span class="dt">col_types =</span> <span class="kw">cols</span>())</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">sim_data <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; # A tibble: 6 x 11</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#&gt;       i    qa    qs    qn     x     n expected_rho subset_qa_qs_qn…</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="co">#&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;           </span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co">#&gt; 1     1 0.428 0.438 0.134 0.274  6390        0.548 S1_0.4279_0.438…</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#&gt; 2     1 0.428 0.438 0.134 0.274  6390        0.548 S1_0.4279_0.438…</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="co">#&gt; 3     1 0.428 0.438 0.134 0.274  6390        0.548 S1_0.4279_0.438…</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#&gt; 4     1 0.428 0.438 0.134 0.274  6390        0.548 S1_0.4279_0.438…</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#&gt; 5     1 0.428 0.438 0.134 0.274  6390        0.548 S1_0.4279_0.438…</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">#&gt; 6     2 0.417 0.322 0.261 0.944  9721        0.721 S2_0.417_0.3223…</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">#&gt; # … with 3 more variables: replicate_id &lt;chr&gt;, successes &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">#&gt; #   observed_rho &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="co">## HDIs for simulated data</span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17">sim_hdi &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="dt">file =</span> <span class="st">"sim_hdi.csv"</span>, <span class="dt">col_types =</span> <span class="kw">cols</span>())</a>
<a class="sourceLine" id="cb8-18" data-line-number="18">sim_hdi <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb8-19" data-line-number="19"><span class="co">#&gt; # A tibble: 6 x 16</span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="co">#&gt;       i subset    qa     qs     qn     x     n n.obs prior.mean obs.min</span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21"><span class="co">#&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb8-22" data-line-number="22"><span class="co">#&gt; 1     1 S1     0.428 0.438  0.134  0.274  6390     5      0.549   0.544</span></a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="co">#&gt; 2     2 S10    0.414 0.204  0.382  0.204  9486     5      0.451   0.449</span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24"><span class="co">#&gt; 3     3 S100   0.772 0.186  0.0416 0.741  5834     5      0.913   0.910</span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"><span class="co">#&gt; 4     4 S1000  0.466 0.0687 0.465  0.354  1484     5      0.483   0.478</span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="co">#&gt; 5     5 S101   0.489 0.450  0.0611 0.782  7549     5      0.837   0.833</span></a>
<a class="sourceLine" id="cb8-27" data-line-number="27"><span class="co">#&gt; 6     6 S102   0.153 0.411  0.436  0.213  3852     5      0.242   0.239</span></a>
<a class="sourceLine" id="cb8-28" data-line-number="28"><span class="co">#&gt; # … with 6 more variables: obs.max &lt;dbl&gt;, posterior.mean &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb8-29" data-line-number="29"><span class="co">#&gt; #   posterior.hdi.low &lt;dbl&gt;, posterior.hdi.high &lt;dbl&gt;, obs.range &lt;dbl&gt;,</span></a>
<a class="sourceLine" id="cb8-30" data-line-number="30"><span class="co">#&gt; #   hdi.range &lt;dbl&gt;</span></a></code></pre></div>
<p>Which we have generated as follows:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">## simulate a sample with N subsets, by randomly assigning values to {q_A, q_S, q_N}, x_S and n</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">n_subsets =<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">sim_sample =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dv">1</span><span class="op">:</span>n_subsets, <span class="cf">function</span>(i) {</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">    q =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dv">3</span>, <span class="dt">min =</span> <span class="dv">0</span>, <span class="dt">max =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">    q =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(q<span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(q), <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">    d =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(</a>
<a class="sourceLine" id="cb9-7" data-line-number="7">        <span class="dt">i =</span> i,</a>
<a class="sourceLine" id="cb9-8" data-line-number="8">        <span class="dt">qa =</span> q[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb9-9" data-line-number="9">        <span class="dt">qs =</span> q[<span class="dv">2</span>],</a>
<a class="sourceLine" id="cb9-10" data-line-number="10">        <span class="dt">qn =</span> q[<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb9-11" data-line-number="11">    )</a>
<a class="sourceLine" id="cb9-12" data-line-number="12">}) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="st">  </span><span class="kw">bind_rows</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">1e-4</span>), <span class="dt">size =</span> n_subsets, <span class="dt">replace =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">n =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="fl">1e3</span>, <span class="fl">1e4</span>, <span class="dv">1</span>), <span class="dt">size =</span> n_subsets, <span class="dt">replace =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">expected_rho =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>((n<span class="op">*</span>qa <span class="op">+</span><span class="st"> </span>n<span class="op">*</span>qs<span class="op">*</span>x) <span class="op">/</span><span class="st"> </span>n, <span class="dv">4</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-17" data-line-number="17"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">subset_qa_qs_qn_x_n =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"S%s_%s_%s_%s_%s_%s"</span>, i, qa, qs, qn, x, n))</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19">sim_sample <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb9-20" data-line-number="20"></a>
<a class="sourceLine" id="cb9-21" data-line-number="21"><span class="co">## simulate r replicates from the sample by drawing counts in each subset, with probabilities {1, x, 0}</span></a>
<a class="sourceLine" id="cb9-22" data-line-number="22">get_counts &lt;-<span class="st"> </span><span class="cf">function</span>(subset_qa_qs_qn_x_n, r) {</a>
<a class="sourceLine" id="cb9-23" data-line-number="23">    s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(subset_qa_qs_qn_x_n) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-24" data-line-number="24"><span class="st">        </span><span class="kw">separate</span>(subset_qa_qs_qn_x_n, <span class="dt">into =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"subset"</span>, <span class="st">"qa"</span>, <span class="st">"qs"</span>, <span class="st">"qn"</span>, <span class="st">"x"</span>, <span class="st">"n"</span>), <span class="dt">sep =</span> <span class="st">"_"</span>)</a>
<a class="sourceLine" id="cb9-25" data-line-number="25">    </a>
<a class="sourceLine" id="cb9-26" data-line-number="26">    counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(</a>
<a class="sourceLine" id="cb9-27" data-line-number="27">      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"always"</span>, <span class="st">"sometimes"</span>, <span class="st">"never"</span>)), s<span class="op">$</span>n, <span class="dt">prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(s<span class="op">$</span>qa, s<span class="op">$</span>qs, s<span class="op">$</span>qn), <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb9-28" data-line-number="28">    always &lt;-<span class="st"> </span>counts[<span class="st">"always"</span>]</a>
<a class="sourceLine" id="cb9-29" data-line-number="29">    sometimes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(r), counts[<span class="st">"sometimes"</span>], <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(s<span class="op">$</span>x))</a>
<a class="sourceLine" id="cb9-30" data-line-number="30">    per_replicate_successes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(</a>
<a class="sourceLine" id="cb9-31" data-line-number="31">        <span class="dt">subset_qa_qs_qn_x_n =</span> subset_qa_qs_qn_x_n,</a>
<a class="sourceLine" id="cb9-32" data-line-number="32">        <span class="dt">replicate_id =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"R"</span>, <span class="dv">1</span><span class="op">:</span>r, <span class="dt">sep =</span> <span class="st">""</span>),</a>
<a class="sourceLine" id="cb9-33" data-line-number="33">        <span class="dt">successes =</span> always <span class="op">+</span><span class="st"> </span>sometimes,</a>
<a class="sourceLine" id="cb9-34" data-line-number="34">        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb9-35" data-line-number="35">    )</a>
<a class="sourceLine" id="cb9-36" data-line-number="36">    per_replicate_successes</a>
<a class="sourceLine" id="cb9-37" data-line-number="37">}</a>
<a class="sourceLine" id="cb9-38" data-line-number="38"></a>
<a class="sourceLine" id="cb9-39" data-line-number="39">sim_replicates &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(sim_sample<span class="op">$</span>subset_qa_qs_qn_x_n, <span class="cf">function</span>(s) <span class="kw">get_counts</span>(s, <span class="dt">r =</span> <span class="dv">5</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-40" data-line-number="40"><span class="st">  </span><span class="kw">bind_rows</span>()</a>
<a class="sourceLine" id="cb9-41" data-line-number="41"></a>
<a class="sourceLine" id="cb9-42" data-line-number="42">sim_replicates <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb9-43" data-line-number="43"></a>
<a class="sourceLine" id="cb9-44" data-line-number="44"><span class="co">## combine the sample and replicate datasets</span></a>
<a class="sourceLine" id="cb9-45" data-line-number="45">sim_data &lt;-<span class="st"> </span>sim_sample <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-46" data-line-number="46"><span class="st">  </span><span class="kw">inner_join</span>(sim_replicates) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-47" data-line-number="47"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">observed_rho =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(successes <span class="op">/</span><span class="st"> </span>n, <span class="dv">4</span>))</a>
<a class="sourceLine" id="cb9-48" data-line-number="48"></a>
<a class="sourceLine" id="cb9-49" data-line-number="49">sim_data <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb9-50" data-line-number="50"><span class="kw">write_csv</span>(sim_data, <span class="dt">path =</span> <span class="st">"sim_data.csv"</span>)</a>
<a class="sourceLine" id="cb9-51" data-line-number="51"></a>
<a class="sourceLine" id="cb9-52" data-line-number="52"><span class="co">## calculate HDIs</span></a>
<a class="sourceLine" id="cb9-53" data-line-number="53">stan_result &lt;-<span class="st"> </span><span class="kw">sample_posterior</span>(<span class="dt">m =</span> sim_data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">Subset =</span> subset_qa_qs_qn_x_n),</a>
<a class="sourceLine" id="cb9-54" data-line-number="54">                                <span class="dt">successes_field =</span> <span class="st">"successes"</span>, <span class="dt">totals_field =</span> <span class="st">"n"</span>)</a>
<a class="sourceLine" id="cb9-55" data-line-number="55">sim_hdi &lt;-<span class="st"> </span><span class="kw">estimate_hdi</span>(<span class="dt">r =</span> stan_result) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb9-56" data-line-number="56"><span class="st">    </span><span class="kw">separate</span>(subset, <span class="dt">into =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"subset"</span>, <span class="st">"qa"</span>, <span class="st">"qs"</span>, <span class="st">"qn"</span>, <span class="st">"x"</span>, <span class="st">"n"</span>), <span class="dt">sep =</span> <span class="st">"_"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-57" data-line-number="57"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">n =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(n)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-58" data-line-number="58"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">obs.range =</span> obs.max <span class="op">-</span><span class="st"> </span>obs.min) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-59" data-line-number="59"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">hdi.range =</span> posterior.hdi.high <span class="op">-</span><span class="st"> </span>posterior.hdi.low)</a>
<a class="sourceLine" id="cb9-60" data-line-number="60"></a>
<a class="sourceLine" id="cb9-61" data-line-number="61">sim_hdi <span class="op">%&gt;%</span><span class="st"> </span>head</a>
<a class="sourceLine" id="cb9-62" data-line-number="62"><span class="kw">write_csv</span>(sim_hdi, <span class="dt">path =</span> <span class="st">"sim_hdi.csv"</span>)</a></code></pre></div>
</div>
<div id="simulation-results" class="section level3">
<h3 class="hasAnchor">
<a href="#simulation-results" class="anchor"></a>Simulation results</h3>
<p>By comparing the expected vs. observed <span class="math inline">\(\rho\)</span> from our simulated dataset, we can confirm that our simulation has indeed produced the intended variability:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">sim_data <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw">ggplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw">geom_abline</span>(<span class="dt">slope =</span> <span class="dv">1</span>, <span class="dt">color =</span> <span class="st">"gray70"</span>, <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="fl">0.25</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x =</span> expected_rho, <span class="dt">y =</span> observed_rho, <span class="dt">color =</span> replicate_id), <span class="dt">shape =</span> <span class="dv">1</span>, <span class="dt">size =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="stratified_counts_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>We can also inspect if our HDIs respond to the variables of interest, that is, subset size and replicate variability:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">p1 &lt;-<span class="st"> </span>sim_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> n, <span class="dt">y =</span> hdi.range)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">"HDI range decreases with subset size"</span>)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"></a>
<a class="sourceLine" id="cb11-6" data-line-number="6">p2 &lt;-<span class="st"> </span>sim_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> obs.range, <span class="dt">y =</span> hdi.range)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">"HDI range increases with replicate variability"</span>)</a>
<a class="sourceLine" id="cb11-10" data-line-number="10"></a>
<a class="sourceLine" id="cb11-11" data-line-number="11">gridExtra<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/gridExtra/topics/arrangeGrob">grid.arrange</a></span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="stratified_counts_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>We can confirm that our HDIs contain the simulated per-replicate probabilities by inspecting a few subsets:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">i &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim</a></span>(sim_hdi)[<span class="dv">1</span>]), <span class="dt">size =</span> <span class="dv">10</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">sel_subsets &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"S"</span>, i, <span class="dt">sep =</span> <span class="st">""</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">dodge_width &lt;-<span class="st"> </span><span class="fl">0.7</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">sim_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(subset <span class="op">%in%</span><span class="st"> </span>sel_subsets) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> subset)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="st">    </span><span class="co"># observed</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> obs.min, <span class="dt">ymax =</span> obs.max), </a>
<a class="sourceLine" id="cb12-10" data-line-number="10">                  <span class="dt">alpha =</span> <span class="fl">0.2</span>, <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width), <span class="dt">width =</span> <span class="dv">0</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="st">    </span><span class="co"># estimated</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> posterior.hdi.low, <span class="dt">ymax =</span> posterior.hdi.high), </a>
<a class="sourceLine" id="cb12-13" data-line-number="13">                  <span class="dt">size =</span> <span class="fl">0.5</span>, <span class="dt">alpha =</span> <span class="dv">1</span>, <span class="dt">width =</span> <span class="fl">0.6</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">y =</span> posterior.mean), <span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">shape =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb12-15" data-line-number="15">               <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">ylim</a></span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-18" data-line-number="18"><span class="st">    </span><span class="kw">coord_flip</span>()</a></code></pre></div>
<p><img src="stratified_counts_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>And we can also extend the evaluation to all subsets and quantify what proportion of HDIs contain all 5 replicate observations:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">obs_in_hdi &lt;-<span class="st"> </span>sim_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">obs_min_in_hdi =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(posterior.hdi.low <span class="op">&lt;=</span><span class="st"> </span>obs.min, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">obs_max_in_hdi =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(posterior.hdi.high <span class="op">&gt;=</span><span class="st"> </span>obs.max, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">all_obs_in_hdi =</span> obs_min_in_hdi <span class="op">&amp;</span><span class="st"> </span>obs_max_in_hdi)</a>
<a class="sourceLine" id="cb13-5" data-line-number="5"></a>
<a class="sourceLine" id="cb13-6" data-line-number="6">obs_in_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="st">    </span><span class="kw">group_by</span>(all_obs_in_hdi) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8"><span class="st">    </span><span class="kw">summarise</span>(<span class="dt">n =</span> <span class="kw">n</span>())</a>
<a class="sourceLine" id="cb13-9" data-line-number="9"><span class="co">#&gt; # A tibble: 2 x 2</span></a>
<a class="sourceLine" id="cb13-10" data-line-number="10"><span class="co">#&gt;   all_obs_in_hdi     n</span></a>
<a class="sourceLine" id="cb13-11" data-line-number="11"><span class="co">#&gt;   &lt;lgl&gt;          &lt;int&gt;</span></a>
<a class="sourceLine" id="cb13-12" data-line-number="12"><span class="co">#&gt; 1 FALSE             74</span></a>
<a class="sourceLine" id="cb13-13" data-line-number="13"><span class="co">#&gt; 2 TRUE             926</span></a></code></pre></div>
<p>Finally, let’s inspect a few cases where our HDIs do not capture all of the observations in our simulation:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">sel_subsets &lt;-<span class="st"> </span>obs_in_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(all_obs_in_hdi <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="st">    </span><span class="kw">select</span>(subset) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>()</a>
<a class="sourceLine" id="cb14-6" data-line-number="6">dodge_width &lt;-<span class="st"> </span><span class="fl">0.7</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"></a>
<a class="sourceLine" id="cb14-8" data-line-number="8">sim_hdi <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(subset <span class="op">%in%</span><span class="st"> </span>sel_subsets) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-10" data-line-number="10"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> subset)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-11" data-line-number="11"><span class="st">    </span><span class="co"># observed</span></a>
<a class="sourceLine" id="cb14-12" data-line-number="12"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> obs.min, <span class="dt">ymax =</span> obs.max), </a>
<a class="sourceLine" id="cb14-13" data-line-number="13">                  <span class="dt">alpha =</span> <span class="fl">0.2</span>, <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width), <span class="dt">width =</span> <span class="dv">0</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-14" data-line-number="14"><span class="st">    </span><span class="co"># estimated</span></a>
<a class="sourceLine" id="cb14-15" data-line-number="15"><span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> posterior.hdi.low, <span class="dt">ymax =</span> posterior.hdi.high), </a>
<a class="sourceLine" id="cb14-16" data-line-number="16">                  <span class="dt">size =</span> <span class="fl">0.5</span>, <span class="dt">alpha =</span> <span class="dv">1</span>, <span class="dt">width =</span> <span class="fl">0.6</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-17" data-line-number="17"><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">y =</span> posterior.mean), <span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">shape =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb14-18" data-line-number="18">               <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> dodge_width)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-19" data-line-number="19"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-20" data-line-number="20"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">ylim</a></span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-21" data-line-number="21"><span class="st">    </span><span class="kw">coord_flip</span>()</a></code></pre></div>
<p><img src="stratified_counts_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#set-up">Set up</a></li>
      <li><a href="#comparing-stratified-counts-across-analysis-groups">Comparing stratified counts across analysis groups</a></li>
      <li><a href="#calibrating-hdis-with-simulated-data">Calibrating HDIs with simulated data</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
