<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract hap.py Precision-Recall data — pr_data • happyR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">happyR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Precision-recall_curves.html">Precision-recall curves</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract hap.py Precision-Recall data</h1>
    </div>

    
    <p>Simpler interface to retrieve a data.frame
of PR metrics from a happy_result object.</p>
    

    <pre class="usage"><span class='fu'>pr_data</span>(<span class='no'>happy_result</span>, <span class='kw'>var_type</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"both"</span>, <span class='st'>"snv"</span>, <span class='st'>"indel"</span>),
  <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"ALL"</span>, <span class='st'>"PASS"</span>, <span class='st'>"SEL"</span>), <span class='kw'>subtype</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"*"</span>, <span class='st'>"C16_PLUS"</span>, <span class='st'>"C1_5"</span>,
  <span class='st'>"C6_15"</span>, <span class='st'>"D16_PLUS"</span>, <span class='st'>"D1_5"</span>, <span class='st'>"D6_15"</span>, <span class='st'>"I16_PLUS"</span>, <span class='st'>"I1_5"</span>, <span class='st'>"I6_15"</span>),
  <span class='kw'>subset</span> <span class='kw'>=</span> <span class='st'>"*"</span>, <span class='kw'>quietly</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>happy_result</th>
      <td><p>a happy result loaded
via <code><a href='read_happy.html'>read_happy</a></code></p></td>
    </tr>
    <tr>
      <th>var_type</th>
      <td><p>subset for either insertions
and deletions <code>"indel"</code>, SNVs <code>"snv"</code>
or keep both</p></td>
    </tr>
    <tr>
      <th>filter</th>
      <td><p>include all records (ALL), only
passing (PASS) or with selective filters applied
(SEL)</p></td>
    </tr>
    <tr>
      <th>subtype</th>
      <td><p>variant subtype of the form <code>[IDC]length_range</code>,
e.g. <code>"D6_15"</code> is deletions of length \(&gt;=5\) and \(&lt;=15\)</p></td>
    </tr>
    <tr>
      <th>subset</th>
      <td><p>when run with stratification regions, the subset is
the region ID. <code>"*"</code> for genome-wide PR data. See details.</p></td>
    </tr>
    <tr>
      <th>quietly</th>
      <td><p>suppress info messages</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Subsets</strong>: hap.py v0.3.7+ writes subsets <code>TS_contained</code> and
<code>TS_boundary</code> by default, corresponding to truth variants
well contained or at the boundary of confident regions. In some
truthsets, those in <code>TS_boundary</code> will show worse performance
metrics due to issues with variant representation or a partial
haplotype description.
    <strong>Subtypes</strong>: Insertion subtypes are of the form: <code>[IDC]length_range</code>
where the first letter indicates the variant classification: <code>I</code> insertion;
<code>D</code> deletion; and <code>C</code> complex. Hap.py bins the lengths of these records
into ranges by ALT allele length in basepairs: <code>1_5</code>, <code>6_15</code> and <code>16_PLUS</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># figure out prefix from pkg install location</span>
<span class='no'>happy_input</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='st'>"happy_demo.summary.csv"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"happyR"</span>)
<span class='no'>happy_prefix</span> <span class='kw'>&lt;-</span> <span class='fu'>sub</span>(<span class='st'>".summary.csv"</span>, <span class='st'>""</span>, <span class='no'>happy_input</span>)

<span class='co'># load happy result</span>
<span class='no'>hapdata</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read_happy.html'>read_happy</a></span>(<span class='no'>happy_prefix</span>)</div><div class='output co'>#&gt; <span class='message'>Reading summary table</span></div><div class='output co'>#&gt; <span class='message'>Reading extended table</span></div><div class='output co'>#&gt; <span class='message'>Reading precision-recall curve data</span></div><div class='input'>
<span class='co'># long deletion PR curve</span>
<span class='no'>del_pr</span> <span class='kw'>&lt;-</span> <span class='fu'>pr_data</span>(<span class='no'>hapdata</span>, <span class='kw'>var_type</span> <span class='kw'>=</span> <span class='st'>"indel"</span>, <span class='kw'>subtype</span> <span class='kw'>=</span> <span class='st'>"D16_PLUS"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Ben Moore, Mar Gonzalez-Porta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
